{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0f92a1a4-2f9b-4bf1-9dbf-fc5de744c23c",
   "metadata": {},
   "source": [
    "# Lab: Gene regulatory network inference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1b4e48-00f9-4312-92cc-065bae54362d",
   "metadata": {},
   "source": [
    "In this lab you will apply a few GRN inference methods to data and compare the resulting GRN predictions with different metrics.\n",
    "\n",
    "The lab consists of the sections:\n",
    "1. Installation\n",
    "2. Obtain the datasets\n",
    "3. GRN inference\n",
    "4. Selecting the number of interactions in inferred GRNs\n",
    "5. Consensus network\n",
    "6. Comparing topological properties\n",
    "7. Comparison with gold standard network\n",
    "8. Plotting GRNs\n",
    "9. Biological interpretation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "698375e1-e3f6-459f-8fa2-c84efb60ac27",
   "metadata": {},
   "source": [
    "## 1. Installation\n",
    "To run this lab you need to have Python 3 and the following packages installed:\n",
    "- Pandas\n",
    "- NumPy\n",
    "- scikit-learn\n",
    "- Matplotlib\n",
    "- seaborn\n",
    "- networkx\n",
    "\n",
    "We have provided a conda environment for this lab which is optional to use for installing the packages for the lab. Download \"environment_grn_lab.yml\" from the course Canvas page and run \"conda env create -f environment_grn_lab.yml\".\n",
    "\n",
    "You should also download the file GENIE3.py from the course Canvas page and place it in your working directory.\n",
    "\n",
    "Start with importing the packages."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b9f678-f53d-418c-bd74-9185a7262b34",
   "metadata": {},
   "source": [
    "## 2. Obtain the datasets\n",
    "\n",
    "For this lab we will use a synthetic dataset which was generated using [GeneNetWeaver](https://academic.oup.com/bioinformatics/article/27/16/2263/254752?login=false). Download the following files from the course Canvas page.\n",
    "- GRN_lab_gene_expression_data.csv - Synthetic gene expression data for 500 genes. The dataset contains 500 samples where one gene has been perturbed in each sample.\n",
    "- GRN_lab_regulators.csv - A list of regulators.\n",
    "- GRN_lab_signed_gold_standard.csv - A gold standard network. This network was used for generating the synthetic gene expression data.\n",
    "\n",
    "Read the gene expression data into a Pandas DataFrame and the list of regulators into a NumPy array by running the code below. Adjust the file paths to your file system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "799fb63a-c5d1-4a6c-a9d8-40572b3395d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of possible target genes: 500\n",
      "Number of samples: 500\n",
      "Number of regulators: 89\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "## Gene expression data\n",
    "exp = pd.read_csv('GRN_lab_gene_expression_data.csv', index_col=0)#.iloc[:, :200]\n",
    "\n",
    "print(f\"Number of possible target genes: {exp.shape[1]}\")\n",
    "print(f\"Number of samples: {exp.shape[0]}\")\n",
    "\n",
    "## Regulators\n",
    "regulators = pd.read_csv('GRN_lab_regulators.csv',\n",
    "                  header=None).values.flatten()\n",
    "print(f\"Number of regulators: {len(regulators)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345a4c83-7375-4bf3-9202-c3703c5c5cdb",
   "metadata": {},
   "source": [
    "## 3. GRN inference\n",
    "In this section you will apply a few GRN inference methods. We have included implementations of three methods: absolute value of the Pearson's correlation coefficient (PCC), [GENIE3](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776), and LASSO. All three methods scores interactions between regulators (often transcription factors; TFs) and their target genes using gene expression data. GENIE3 and LASSO are both regression based method using different types of feature selection. GENIE3 is a tree-based ensemble method and LASSO is a type of regularized linear regression. \n",
    "\n",
    "To your help you we have have provided the following functions:\n",
    "- run_pcc - returns the absolute value of the PCC for each regulator-target gene pair.\n",
    "- run_genie3 - returns a confidence score for each regulator-target gene pair.\n",
    "- run_lasso - returns the abosulute value of the estimated coefficients for each regulator-target gene pair.\n",
    "\n",
    "The output of all functions are a confidence score matrix for interactions between regulators (rows) and target genes (columns). For this lab we consider the metrics obtained from each GRN inference methods to be a confidence score for each regulator-target gene interaction i.e. how likely the interaction is to occur. \n",
    "\n",
    "Start with running the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "878e9437-f16c-446e-93bb-d8f3ea471051",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Absolute value of the PCC\n",
    "def run_pcc(exp, tfs):\n",
    "    # Calculate correlations\n",
    "    corrmat = exp.corr().abs()\n",
    "    confidence_scores = corrmat.loc[tfs, :]\n",
    "    return confidence_scores\n",
    "\n",
    "# GENIE3\n",
    "def run_genie3(exp, tfs, ntrees=100):\n",
    "    # ntrees are set to 100 to reduce computation time.\n",
    "    # Import GENIE3 functions. GENIE3.py file need to be placed in working directory.\n",
    "    from GENIE3 import GENIE3\n",
    "\n",
    "    # Run GENIE3\n",
    "    vim = GENIE3(exp.values, \n",
    "                 gene_names=exp.columns.values.tolist(), \n",
    "                 regulators=tfs.tolist(), \n",
    "                 ntrees=ntrees)\n",
    "    confidence_scores = pd.DataFrame(vim, index=exp.columns, columns=exp.columns).loc[tfs, :]\n",
    "    return confidence_scores\n",
    "\n",
    "# LASSO\n",
    "def run_lasso(exp, tfs, cv=3):\n",
    "    # Import LASSO. Stop warnings from being printed.\n",
    "    import warnings\n",
    "    warnings.filterwarnings(\"ignore\")\n",
    "    from sklearn.linear_model import LassoCV\n",
    "\n",
    "    # Format data\n",
    "    targets_exp = exp.T.values\n",
    "    tfs_exp = exp.loc[:, tfs].values\n",
    "\n",
    "    # Run LASSO with cross validation to select best model parameters.\n",
    "    model = LassoCV(n_jobs=-1, cv=cv)\n",
    "\n",
    "    # Fit one model for each target gene.\n",
    "    coef_mat = np.empty((targets_exp.shape[0], tfs_exp.shape[1]))\n",
    "    for i, j in enumerate(targets_exp):\n",
    "        if i%100 == 0:\n",
    "            print('Working on target number', i)\n",
    "        model.fit(tfs_exp, targets_exp[i])\n",
    "        coef_mat[i] = model.coef_\n",
    "\n",
    "    # Format confidence score matrix and return the absolute value of all coefficients.\n",
    "    confidence_scores = pd.DataFrame(coef_mat, index=exp.columns, columns=tfs).T.abs()\n",
    "    return confidence_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d39d37bb-b30d-4927-bf9e-5475e7e83cdc",
   "metadata": {},
   "source": [
    "Now you can run the GRN inference methods. An easy way to store the inferred GRNs for subsequent processing is to use a dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "633b71f9-e824-43c2-af3d-c1323b156b92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hipA</th>\n",
       "      <th>hipB</th>\n",
       "      <th>ihfB</th>\n",
       "      <th>yeiL</th>\n",
       "      <th>tdcA</th>\n",
       "      <th>tdcR</th>\n",
       "      <th>yiaJ</th>\n",
       "      <th>ompR</th>\n",
       "      <th>bolA</th>\n",
       "      <th>csgD</th>\n",
       "      <th>...</th>\n",
       "      <th>lldP</th>\n",
       "      <th>rfaY</th>\n",
       "      <th>betB</th>\n",
       "      <th>fliT</th>\n",
       "      <th>rpsS</th>\n",
       "      <th>fucK</th>\n",
       "      <th>rpmC</th>\n",
       "      <th>wzb</th>\n",
       "      <th>nemA</th>\n",
       "      <th>prpD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arcA</th>\n",
       "      <td>0.080573</td>\n",
       "      <td>0.015678</td>\n",
       "      <td>0.031990</td>\n",
       "      <td>0.088718</td>\n",
       "      <td>0.039299</td>\n",
       "      <td>0.034481</td>\n",
       "      <td>0.010955</td>\n",
       "      <td>0.033238</td>\n",
       "      <td>0.006733</td>\n",
       "      <td>0.003185</td>\n",
       "      <td>...</td>\n",
       "      <td>0.305528</td>\n",
       "      <td>0.009476</td>\n",
       "      <td>0.417128</td>\n",
       "      <td>0.048657</td>\n",
       "      <td>0.141017</td>\n",
       "      <td>0.003390</td>\n",
       "      <td>0.056532</td>\n",
       "      <td>0.012015</td>\n",
       "      <td>0.032632</td>\n",
       "      <td>0.080441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ascG</th>\n",
       "      <td>0.033339</td>\n",
       "      <td>0.011857</td>\n",
       "      <td>0.037420</td>\n",
       "      <td>0.040001</td>\n",
       "      <td>0.019709</td>\n",
       "      <td>0.038613</td>\n",
       "      <td>0.009065</td>\n",
       "      <td>0.026632</td>\n",
       "      <td>0.054765</td>\n",
       "      <td>0.015743</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041201</td>\n",
       "      <td>0.019870</td>\n",
       "      <td>0.005459</td>\n",
       "      <td>0.051628</td>\n",
       "      <td>0.065479</td>\n",
       "      <td>0.045676</td>\n",
       "      <td>0.013885</td>\n",
       "      <td>0.022122</td>\n",
       "      <td>0.029069</td>\n",
       "      <td>0.003807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>asnC</th>\n",
       "      <td>0.004092</td>\n",
       "      <td>0.045441</td>\n",
       "      <td>0.023317</td>\n",
       "      <td>0.090187</td>\n",
       "      <td>0.019635</td>\n",
       "      <td>0.046327</td>\n",
       "      <td>0.009678</td>\n",
       "      <td>0.050300</td>\n",
       "      <td>0.037683</td>\n",
       "      <td>0.045165</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.036366</td>\n",
       "      <td>0.003173</td>\n",
       "      <td>0.034637</td>\n",
       "      <td>0.015324</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.030390</td>\n",
       "      <td>0.019883</td>\n",
       "      <td>0.031771</td>\n",
       "      <td>0.055216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>baeR</th>\n",
       "      <td>0.017692</td>\n",
       "      <td>0.036308</td>\n",
       "      <td>0.062662</td>\n",
       "      <td>0.007058</td>\n",
       "      <td>0.027523</td>\n",
       "      <td>0.019183</td>\n",
       "      <td>0.012178</td>\n",
       "      <td>0.014862</td>\n",
       "      <td>0.030696</td>\n",
       "      <td>0.033264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033716</td>\n",
       "      <td>0.003347</td>\n",
       "      <td>0.079104</td>\n",
       "      <td>0.078120</td>\n",
       "      <td>0.048729</td>\n",
       "      <td>0.003616</td>\n",
       "      <td>0.083982</td>\n",
       "      <td>0.031858</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.041294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>betI</th>\n",
       "      <td>0.038636</td>\n",
       "      <td>0.040155</td>\n",
       "      <td>0.001730</td>\n",
       "      <td>0.055885</td>\n",
       "      <td>0.058349</td>\n",
       "      <td>0.038609</td>\n",
       "      <td>0.019643</td>\n",
       "      <td>0.038149</td>\n",
       "      <td>0.069182</td>\n",
       "      <td>0.054924</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014714</td>\n",
       "      <td>0.049471</td>\n",
       "      <td>0.272135</td>\n",
       "      <td>0.013217</td>\n",
       "      <td>0.037440</td>\n",
       "      <td>0.026143</td>\n",
       "      <td>0.183229</td>\n",
       "      <td>0.007208</td>\n",
       "      <td>0.028660</td>\n",
       "      <td>0.031904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uxuR</th>\n",
       "      <td>0.017567</td>\n",
       "      <td>0.047054</td>\n",
       "      <td>0.018369</td>\n",
       "      <td>0.047364</td>\n",
       "      <td>0.063203</td>\n",
       "      <td>0.015162</td>\n",
       "      <td>0.191878</td>\n",
       "      <td>0.053587</td>\n",
       "      <td>0.076380</td>\n",
       "      <td>0.007130</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056625</td>\n",
       "      <td>0.059142</td>\n",
       "      <td>0.011379</td>\n",
       "      <td>0.008829</td>\n",
       "      <td>0.032455</td>\n",
       "      <td>0.219120</td>\n",
       "      <td>0.015274</td>\n",
       "      <td>0.031267</td>\n",
       "      <td>0.058980</td>\n",
       "      <td>0.167558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xylR</th>\n",
       "      <td>0.014782</td>\n",
       "      <td>0.068917</td>\n",
       "      <td>0.022958</td>\n",
       "      <td>0.145854</td>\n",
       "      <td>0.082871</td>\n",
       "      <td>0.045244</td>\n",
       "      <td>0.138501</td>\n",
       "      <td>0.175845</td>\n",
       "      <td>0.080273</td>\n",
       "      <td>0.035463</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051127</td>\n",
       "      <td>0.036481</td>\n",
       "      <td>0.056773</td>\n",
       "      <td>0.088306</td>\n",
       "      <td>0.051249</td>\n",
       "      <td>0.141203</td>\n",
       "      <td>0.004518</td>\n",
       "      <td>0.032938</td>\n",
       "      <td>0.029679</td>\n",
       "      <td>0.205164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydeO</th>\n",
       "      <td>0.002158</td>\n",
       "      <td>0.045595</td>\n",
       "      <td>0.011227</td>\n",
       "      <td>0.039420</td>\n",
       "      <td>0.067356</td>\n",
       "      <td>0.005965</td>\n",
       "      <td>0.060683</td>\n",
       "      <td>0.032746</td>\n",
       "      <td>0.021724</td>\n",
       "      <td>0.011689</td>\n",
       "      <td>...</td>\n",
       "      <td>0.091300</td>\n",
       "      <td>0.037146</td>\n",
       "      <td>0.038097</td>\n",
       "      <td>0.020529</td>\n",
       "      <td>0.030810</td>\n",
       "      <td>0.076725</td>\n",
       "      <td>0.046496</td>\n",
       "      <td>0.094874</td>\n",
       "      <td>0.029798</td>\n",
       "      <td>0.012072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yiaJ</th>\n",
       "      <td>0.026414</td>\n",
       "      <td>0.008783</td>\n",
       "      <td>0.039934</td>\n",
       "      <td>0.075326</td>\n",
       "      <td>0.108021</td>\n",
       "      <td>0.028481</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.252763</td>\n",
       "      <td>0.118462</td>\n",
       "      <td>0.022106</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019994</td>\n",
       "      <td>0.076873</td>\n",
       "      <td>0.002565</td>\n",
       "      <td>0.099403</td>\n",
       "      <td>0.039337</td>\n",
       "      <td>0.204927</td>\n",
       "      <td>0.025341</td>\n",
       "      <td>0.013933</td>\n",
       "      <td>0.038584</td>\n",
       "      <td>0.154237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zraR</th>\n",
       "      <td>0.145557</td>\n",
       "      <td>0.094194</td>\n",
       "      <td>0.013594</td>\n",
       "      <td>0.124162</td>\n",
       "      <td>0.043722</td>\n",
       "      <td>0.011673</td>\n",
       "      <td>0.294884</td>\n",
       "      <td>0.226214</td>\n",
       "      <td>0.085994</td>\n",
       "      <td>0.080594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010081</td>\n",
       "      <td>0.051013</td>\n",
       "      <td>0.128140</td>\n",
       "      <td>0.020947</td>\n",
       "      <td>0.047163</td>\n",
       "      <td>0.247801</td>\n",
       "      <td>0.055854</td>\n",
       "      <td>0.059795</td>\n",
       "      <td>0.035693</td>\n",
       "      <td>0.377308</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 500 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hipA      hipB      ihfB      yeiL      tdcA      tdcR      yiaJ  \\\n",
       "arcA  0.080573  0.015678  0.031990  0.088718  0.039299  0.034481  0.010955   \n",
       "ascG  0.033339  0.011857  0.037420  0.040001  0.019709  0.038613  0.009065   \n",
       "asnC  0.004092  0.045441  0.023317  0.090187  0.019635  0.046327  0.009678   \n",
       "baeR  0.017692  0.036308  0.062662  0.007058  0.027523  0.019183  0.012178   \n",
       "betI  0.038636  0.040155  0.001730  0.055885  0.058349  0.038609  0.019643   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "uxuR  0.017567  0.047054  0.018369  0.047364  0.063203  0.015162  0.191878   \n",
       "xylR  0.014782  0.068917  0.022958  0.145854  0.082871  0.045244  0.138501   \n",
       "ydeO  0.002158  0.045595  0.011227  0.039420  0.067356  0.005965  0.060683   \n",
       "yiaJ  0.026414  0.008783  0.039934  0.075326  0.108021  0.028481  1.000000   \n",
       "zraR  0.145557  0.094194  0.013594  0.124162  0.043722  0.011673  0.294884   \n",
       "\n",
       "          ompR      bolA      csgD  ...      lldP      rfaY      betB  \\\n",
       "arcA  0.033238  0.006733  0.003185  ...  0.305528  0.009476  0.417128   \n",
       "ascG  0.026632  0.054765  0.015743  ...  0.041201  0.019870  0.005459   \n",
       "asnC  0.050300  0.037683  0.045165  ...  0.000006  0.036366  0.003173   \n",
       "baeR  0.014862  0.030696  0.033264  ...  0.033716  0.003347  0.079104   \n",
       "betI  0.038149  0.069182  0.054924  ...  0.014714  0.049471  0.272135   \n",
       "...        ...       ...       ...  ...       ...       ...       ...   \n",
       "uxuR  0.053587  0.076380  0.007130  ...  0.056625  0.059142  0.011379   \n",
       "xylR  0.175845  0.080273  0.035463  ...  0.051127  0.036481  0.056773   \n",
       "ydeO  0.032746  0.021724  0.011689  ...  0.091300  0.037146  0.038097   \n",
       "yiaJ  0.252763  0.118462  0.022106  ...  0.019994  0.076873  0.002565   \n",
       "zraR  0.226214  0.085994  0.080594  ...  0.010081  0.051013  0.128140   \n",
       "\n",
       "          fliT      rpsS      fucK      rpmC       wzb      nemA      prpD  \n",
       "arcA  0.048657  0.141017  0.003390  0.056532  0.012015  0.032632  0.080441  \n",
       "ascG  0.051628  0.065479  0.045676  0.013885  0.022122  0.029069  0.003807  \n",
       "asnC  0.034637  0.015324  0.000411  0.030390  0.019883  0.031771  0.055216  \n",
       "baeR  0.078120  0.048729  0.003616  0.083982  0.031858  0.000677  0.041294  \n",
       "betI  0.013217  0.037440  0.026143  0.183229  0.007208  0.028660  0.031904  \n",
       "...        ...       ...       ...       ...       ...       ...       ...  \n",
       "uxuR  0.008829  0.032455  0.219120  0.015274  0.031267  0.058980  0.167558  \n",
       "xylR  0.088306  0.051249  0.141203  0.004518  0.032938  0.029679  0.205164  \n",
       "ydeO  0.020529  0.030810  0.076725  0.046496  0.094874  0.029798  0.012072  \n",
       "yiaJ  0.099403  0.039337  0.204927  0.025341  0.013933  0.038584  0.154237  \n",
       "zraR  0.020947  0.047163  0.247801  0.055854  0.059795  0.035693  0.377308  \n",
       "\n",
       "[89 rows x 500 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run PCC\n",
    "grns_scores = {}\n",
    "grns_scores['PCC'] = run_pcc(exp, regulators)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1188c4b5-f874-4fa5-a8c5-989afee1ea66",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:71: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if regulators is not 'all':\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:82: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if maxcount is not 'all' and not isinstance(maxcount,int):\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:219: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if regulators is not 'all':\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:230: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if tree_method is not 'RF' and tree_method is not 'ET':\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:230: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if tree_method is not 'RF' and tree_method is not 'ET':\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:233: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if K is not 'sqrt' and K is not 'all' and not isinstance(K,int):\n",
      "/mnt/data/hong/courses/SystemBiology/GRN/GENIE3.py:233: SyntaxWarning: \"is not\" with 'str' literal. Did you mean \"!=\"?\n",
      "  if K is not 'sqrt' and K is not 'all' and not isinstance(K,int):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree method: RF\n",
      "K: sqrt\n",
      "Number of trees: 100\n",
      "\n",
      "\n",
      "running single threaded jobs\n",
      "Gene 1/500...\n",
      "Gene 2/500...\n",
      "Gene 3/500...\n",
      "Gene 4/500...\n",
      "Gene 5/500...\n",
      "Gene 6/500...\n",
      "Gene 7/500...\n",
      "Gene 8/500...\n",
      "Gene 9/500...\n",
      "Gene 10/500...\n",
      "Gene 11/500...\n",
      "Gene 12/500...\n",
      "Gene 13/500...\n",
      "Gene 14/500...\n",
      "Gene 15/500...\n",
      "Gene 16/500...\n",
      "Gene 17/500...\n",
      "Gene 18/500...\n",
      "Gene 19/500...\n",
      "Gene 20/500...\n",
      "Gene 21/500...\n",
      "Gene 22/500...\n",
      "Gene 23/500...\n",
      "Gene 24/500...\n",
      "Gene 25/500...\n",
      "Gene 26/500...\n",
      "Gene 27/500...\n",
      "Gene 28/500...\n",
      "Gene 29/500...\n",
      "Gene 30/500...\n",
      "Gene 31/500...\n",
      "Gene 32/500...\n",
      "Gene 33/500...\n",
      "Gene 34/500...\n",
      "Gene 35/500...\n",
      "Gene 36/500...\n",
      "Gene 37/500...\n",
      "Gene 38/500...\n",
      "Gene 39/500...\n",
      "Gene 40/500...\n",
      "Gene 41/500...\n",
      "Gene 42/500...\n",
      "Gene 43/500...\n",
      "Gene 44/500...\n",
      "Gene 45/500...\n",
      "Gene 46/500...\n",
      "Gene 47/500...\n",
      "Gene 48/500...\n",
      "Gene 49/500...\n",
      "Gene 50/500...\n",
      "Gene 51/500...\n",
      "Gene 52/500...\n",
      "Gene 53/500...\n",
      "Gene 54/500...\n",
      "Gene 55/500...\n",
      "Gene 56/500...\n",
      "Gene 57/500...\n",
      "Gene 58/500...\n",
      "Gene 59/500...\n",
      "Gene 60/500...\n",
      "Gene 61/500...\n",
      "Gene 62/500...\n",
      "Gene 63/500...\n",
      "Gene 64/500...\n",
      "Gene 65/500...\n",
      "Gene 66/500...\n",
      "Gene 67/500...\n",
      "Gene 68/500...\n",
      "Gene 69/500...\n",
      "Gene 70/500...\n",
      "Gene 71/500...\n",
      "Gene 72/500...\n",
      "Gene 73/500...\n",
      "Gene 74/500...\n",
      "Gene 75/500...\n",
      "Gene 76/500...\n",
      "Gene 77/500...\n",
      "Gene 78/500...\n",
      "Gene 79/500...\n",
      "Gene 80/500...\n",
      "Gene 81/500...\n",
      "Gene 82/500...\n",
      "Gene 83/500...\n",
      "Gene 84/500...\n",
      "Gene 85/500...\n",
      "Gene 86/500...\n",
      "Gene 87/500...\n",
      "Gene 88/500...\n",
      "Gene 89/500...\n",
      "Gene 90/500...\n",
      "Gene 91/500...\n",
      "Gene 92/500...\n",
      "Gene 93/500...\n",
      "Gene 94/500...\n",
      "Gene 95/500...\n",
      "Gene 96/500...\n",
      "Gene 97/500...\n",
      "Gene 98/500...\n",
      "Gene 99/500...\n",
      "Gene 100/500...\n",
      "Gene 101/500...\n",
      "Gene 102/500...\n",
      "Gene 103/500...\n",
      "Gene 104/500...\n",
      "Gene 105/500...\n",
      "Gene 106/500...\n",
      "Gene 107/500...\n",
      "Gene 108/500...\n",
      "Gene 109/500...\n",
      "Gene 110/500...\n",
      "Gene 111/500...\n",
      "Gene 112/500...\n",
      "Gene 113/500...\n",
      "Gene 114/500...\n",
      "Gene 115/500...\n",
      "Gene 116/500...\n",
      "Gene 117/500...\n",
      "Gene 118/500...\n",
      "Gene 119/500...\n",
      "Gene 120/500...\n",
      "Gene 121/500...\n",
      "Gene 122/500...\n",
      "Gene 123/500...\n",
      "Gene 124/500...\n",
      "Gene 125/500...\n",
      "Gene 126/500...\n",
      "Gene 127/500...\n",
      "Gene 128/500...\n",
      "Gene 129/500...\n",
      "Gene 130/500...\n",
      "Gene 131/500...\n",
      "Gene 132/500...\n",
      "Gene 133/500...\n",
      "Gene 134/500...\n",
      "Gene 135/500...\n",
      "Gene 136/500...\n",
      "Gene 137/500...\n",
      "Gene 138/500...\n",
      "Gene 139/500...\n",
      "Gene 140/500...\n",
      "Gene 141/500...\n",
      "Gene 142/500...\n",
      "Gene 143/500...\n",
      "Gene 144/500...\n",
      "Gene 145/500...\n",
      "Gene 146/500...\n",
      "Gene 147/500...\n",
      "Gene 148/500...\n",
      "Gene 149/500...\n",
      "Gene 150/500...\n",
      "Gene 151/500...\n",
      "Gene 152/500...\n",
      "Gene 153/500...\n",
      "Gene 154/500...\n",
      "Gene 155/500...\n",
      "Gene 156/500...\n",
      "Gene 157/500...\n",
      "Gene 158/500...\n",
      "Gene 159/500...\n",
      "Gene 160/500...\n",
      "Gene 161/500...\n",
      "Gene 162/500...\n",
      "Gene 163/500...\n",
      "Gene 164/500...\n",
      "Gene 165/500...\n",
      "Gene 166/500...\n",
      "Gene 167/500...\n",
      "Gene 168/500...\n",
      "Gene 169/500...\n",
      "Gene 170/500...\n",
      "Gene 171/500...\n",
      "Gene 172/500...\n",
      "Gene 173/500...\n",
      "Gene 174/500...\n",
      "Gene 175/500...\n",
      "Gene 176/500...\n",
      "Gene 177/500...\n",
      "Gene 178/500...\n",
      "Gene 179/500...\n",
      "Gene 180/500...\n",
      "Gene 181/500...\n",
      "Gene 182/500...\n",
      "Gene 183/500...\n",
      "Gene 184/500...\n",
      "Gene 185/500...\n",
      "Gene 186/500...\n",
      "Gene 187/500...\n",
      "Gene 188/500...\n",
      "Gene 189/500...\n",
      "Gene 190/500...\n",
      "Gene 191/500...\n",
      "Gene 192/500...\n",
      "Gene 193/500...\n",
      "Gene 194/500...\n",
      "Gene 195/500...\n",
      "Gene 196/500...\n",
      "Gene 197/500...\n",
      "Gene 198/500...\n",
      "Gene 199/500...\n",
      "Gene 200/500...\n",
      "Gene 201/500...\n",
      "Gene 202/500...\n",
      "Gene 203/500...\n",
      "Gene 204/500...\n",
      "Gene 205/500...\n",
      "Gene 206/500...\n",
      "Gene 207/500...\n",
      "Gene 208/500...\n",
      "Gene 209/500...\n",
      "Gene 210/500...\n",
      "Gene 211/500...\n",
      "Gene 212/500...\n",
      "Gene 213/500...\n",
      "Gene 214/500...\n",
      "Gene 215/500...\n",
      "Gene 216/500...\n",
      "Gene 217/500...\n",
      "Gene 218/500...\n",
      "Gene 219/500...\n",
      "Gene 220/500...\n",
      "Gene 221/500...\n",
      "Gene 222/500...\n",
      "Gene 223/500...\n",
      "Gene 224/500...\n",
      "Gene 225/500...\n",
      "Gene 226/500...\n",
      "Gene 227/500...\n",
      "Gene 228/500...\n",
      "Gene 229/500...\n",
      "Gene 230/500...\n",
      "Gene 231/500...\n",
      "Gene 232/500...\n",
      "Gene 233/500...\n",
      "Gene 234/500...\n",
      "Gene 235/500...\n",
      "Gene 236/500...\n",
      "Gene 237/500...\n",
      "Gene 238/500...\n",
      "Gene 239/500...\n",
      "Gene 240/500...\n",
      "Gene 241/500...\n",
      "Gene 242/500...\n",
      "Gene 243/500...\n",
      "Gene 244/500...\n",
      "Gene 245/500...\n",
      "Gene 246/500...\n",
      "Gene 247/500...\n",
      "Gene 248/500...\n",
      "Gene 249/500...\n",
      "Gene 250/500...\n",
      "Gene 251/500...\n",
      "Gene 252/500...\n",
      "Gene 253/500...\n",
      "Gene 254/500...\n",
      "Gene 255/500...\n",
      "Gene 256/500...\n",
      "Gene 257/500...\n",
      "Gene 258/500...\n",
      "Gene 259/500...\n",
      "Gene 260/500...\n",
      "Gene 261/500...\n",
      "Gene 262/500...\n",
      "Gene 263/500...\n",
      "Gene 264/500...\n",
      "Gene 265/500...\n",
      "Gene 266/500...\n",
      "Gene 267/500...\n",
      "Gene 268/500...\n",
      "Gene 269/500...\n",
      "Gene 270/500...\n",
      "Gene 271/500...\n",
      "Gene 272/500...\n",
      "Gene 273/500...\n",
      "Gene 274/500...\n",
      "Gene 275/500...\n",
      "Gene 276/500...\n",
      "Gene 277/500...\n",
      "Gene 278/500...\n",
      "Gene 279/500...\n",
      "Gene 280/500...\n",
      "Gene 281/500...\n",
      "Gene 282/500...\n",
      "Gene 283/500...\n",
      "Gene 284/500...\n",
      "Gene 285/500...\n",
      "Gene 286/500...\n",
      "Gene 287/500...\n",
      "Gene 288/500...\n",
      "Gene 289/500...\n",
      "Gene 290/500...\n",
      "Gene 291/500...\n",
      "Gene 292/500...\n",
      "Gene 293/500...\n",
      "Gene 294/500...\n",
      "Gene 295/500...\n",
      "Gene 296/500...\n",
      "Gene 297/500...\n",
      "Gene 298/500...\n",
      "Gene 299/500...\n",
      "Gene 300/500...\n",
      "Gene 301/500...\n",
      "Gene 302/500...\n",
      "Gene 303/500...\n",
      "Gene 304/500...\n",
      "Gene 305/500...\n",
      "Gene 306/500...\n",
      "Gene 307/500...\n",
      "Gene 308/500...\n",
      "Gene 309/500...\n",
      "Gene 310/500...\n",
      "Gene 311/500...\n",
      "Gene 312/500...\n",
      "Gene 313/500...\n",
      "Gene 314/500...\n",
      "Gene 315/500...\n",
      "Gene 316/500...\n",
      "Gene 317/500...\n",
      "Gene 318/500...\n",
      "Gene 319/500...\n",
      "Gene 320/500...\n",
      "Gene 321/500...\n",
      "Gene 322/500...\n",
      "Gene 323/500...\n",
      "Gene 324/500...\n",
      "Gene 325/500...\n",
      "Gene 326/500...\n",
      "Gene 327/500...\n",
      "Gene 328/500...\n",
      "Gene 329/500...\n",
      "Gene 330/500...\n",
      "Gene 331/500...\n",
      "Gene 332/500...\n",
      "Gene 333/500...\n",
      "Gene 334/500...\n",
      "Gene 335/500...\n",
      "Gene 336/500...\n",
      "Gene 337/500...\n",
      "Gene 338/500...\n",
      "Gene 339/500...\n",
      "Gene 340/500...\n",
      "Gene 341/500...\n",
      "Gene 342/500...\n",
      "Gene 343/500...\n",
      "Gene 344/500...\n",
      "Gene 345/500...\n",
      "Gene 346/500...\n",
      "Gene 347/500...\n",
      "Gene 348/500...\n",
      "Gene 349/500...\n",
      "Gene 350/500...\n",
      "Gene 351/500...\n",
      "Gene 352/500...\n",
      "Gene 353/500...\n",
      "Gene 354/500...\n",
      "Gene 355/500...\n",
      "Gene 356/500...\n",
      "Gene 357/500...\n",
      "Gene 358/500...\n",
      "Gene 359/500...\n",
      "Gene 360/500...\n",
      "Gene 361/500...\n",
      "Gene 362/500...\n",
      "Gene 363/500...\n",
      "Gene 364/500...\n",
      "Gene 365/500...\n",
      "Gene 366/500...\n",
      "Gene 367/500...\n",
      "Gene 368/500...\n",
      "Gene 369/500...\n",
      "Gene 370/500...\n",
      "Gene 371/500...\n",
      "Gene 372/500...\n",
      "Gene 373/500...\n",
      "Gene 374/500...\n",
      "Gene 375/500...\n",
      "Gene 376/500...\n",
      "Gene 377/500...\n",
      "Gene 378/500...\n",
      "Gene 379/500...\n",
      "Gene 380/500...\n",
      "Gene 381/500...\n",
      "Gene 382/500...\n",
      "Gene 383/500...\n",
      "Gene 384/500...\n",
      "Gene 385/500...\n",
      "Gene 386/500...\n",
      "Gene 387/500...\n",
      "Gene 388/500...\n",
      "Gene 389/500...\n",
      "Gene 390/500...\n",
      "Gene 391/500...\n",
      "Gene 392/500...\n",
      "Gene 393/500...\n",
      "Gene 394/500...\n",
      "Gene 395/500...\n",
      "Gene 396/500...\n",
      "Gene 397/500...\n",
      "Gene 398/500...\n",
      "Gene 399/500...\n",
      "Gene 400/500...\n",
      "Gene 401/500...\n",
      "Gene 402/500...\n",
      "Gene 403/500...\n",
      "Gene 404/500...\n",
      "Gene 405/500...\n",
      "Gene 406/500...\n",
      "Gene 407/500...\n",
      "Gene 408/500...\n",
      "Gene 409/500...\n",
      "Gene 410/500...\n",
      "Gene 411/500...\n",
      "Gene 412/500...\n",
      "Gene 413/500...\n",
      "Gene 414/500...\n",
      "Gene 415/500...\n",
      "Gene 416/500...\n",
      "Gene 417/500...\n",
      "Gene 418/500...\n",
      "Gene 419/500...\n",
      "Gene 420/500...\n",
      "Gene 421/500...\n",
      "Gene 422/500...\n",
      "Gene 423/500...\n",
      "Gene 424/500...\n",
      "Gene 425/500...\n",
      "Gene 426/500...\n",
      "Gene 427/500...\n",
      "Gene 428/500...\n",
      "Gene 429/500...\n",
      "Gene 430/500...\n",
      "Gene 431/500...\n",
      "Gene 432/500...\n",
      "Gene 433/500...\n",
      "Gene 434/500...\n",
      "Gene 435/500...\n",
      "Gene 436/500...\n",
      "Gene 437/500...\n",
      "Gene 438/500...\n",
      "Gene 439/500...\n",
      "Gene 440/500...\n",
      "Gene 441/500...\n",
      "Gene 442/500...\n",
      "Gene 443/500...\n",
      "Gene 444/500...\n",
      "Gene 445/500...\n",
      "Gene 446/500...\n",
      "Gene 447/500...\n",
      "Gene 448/500...\n",
      "Gene 449/500...\n",
      "Gene 450/500...\n",
      "Gene 451/500...\n",
      "Gene 452/500...\n",
      "Gene 453/500...\n",
      "Gene 454/500...\n",
      "Gene 455/500...\n",
      "Gene 456/500...\n",
      "Gene 457/500...\n",
      "Gene 458/500...\n",
      "Gene 459/500...\n",
      "Gene 460/500...\n",
      "Gene 461/500...\n",
      "Gene 462/500...\n",
      "Gene 463/500...\n",
      "Gene 464/500...\n",
      "Gene 465/500...\n",
      "Gene 466/500...\n",
      "Gene 467/500...\n",
      "Gene 468/500...\n",
      "Gene 469/500...\n",
      "Gene 470/500...\n",
      "Gene 471/500...\n",
      "Gene 472/500...\n",
      "Gene 473/500...\n",
      "Gene 474/500...\n",
      "Gene 475/500...\n",
      "Gene 476/500...\n",
      "Gene 477/500...\n",
      "Gene 478/500...\n",
      "Gene 479/500...\n",
      "Gene 480/500...\n",
      "Gene 481/500...\n",
      "Gene 482/500...\n",
      "Gene 483/500...\n",
      "Gene 484/500...\n",
      "Gene 485/500...\n",
      "Gene 486/500...\n",
      "Gene 487/500...\n",
      "Gene 488/500...\n",
      "Gene 489/500...\n",
      "Gene 490/500...\n",
      "Gene 491/500...\n",
      "Gene 492/500...\n",
      "Gene 493/500...\n",
      "Gene 494/500...\n",
      "Gene 495/500...\n",
      "Gene 496/500...\n",
      "Gene 497/500...\n",
      "Gene 498/500...\n",
      "Gene 499/500...\n",
      "Gene 500/500...\n",
      "Elapsed time: 152.10 seconds\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hipA</th>\n",
       "      <th>hipB</th>\n",
       "      <th>ihfB</th>\n",
       "      <th>yeiL</th>\n",
       "      <th>tdcA</th>\n",
       "      <th>tdcR</th>\n",
       "      <th>yiaJ</th>\n",
       "      <th>ompR</th>\n",
       "      <th>bolA</th>\n",
       "      <th>csgD</th>\n",
       "      <th>...</th>\n",
       "      <th>lldP</th>\n",
       "      <th>rfaY</th>\n",
       "      <th>betB</th>\n",
       "      <th>fliT</th>\n",
       "      <th>rpsS</th>\n",
       "      <th>fucK</th>\n",
       "      <th>rpmC</th>\n",
       "      <th>wzb</th>\n",
       "      <th>nemA</th>\n",
       "      <th>prpD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arcA</th>\n",
       "      <td>0.012169</td>\n",
       "      <td>0.012331</td>\n",
       "      <td>0.010293</td>\n",
       "      <td>0.010130</td>\n",
       "      <td>0.009825</td>\n",
       "      <td>0.015032</td>\n",
       "      <td>0.010608</td>\n",
       "      <td>0.009096</td>\n",
       "      <td>0.005641</td>\n",
       "      <td>0.012541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061884</td>\n",
       "      <td>0.009760</td>\n",
       "      <td>0.060858</td>\n",
       "      <td>0.008531</td>\n",
       "      <td>0.020052</td>\n",
       "      <td>0.005519</td>\n",
       "      <td>0.010090</td>\n",
       "      <td>0.008549</td>\n",
       "      <td>0.011201</td>\n",
       "      <td>0.006739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ascG</th>\n",
       "      <td>0.010982</td>\n",
       "      <td>0.009748</td>\n",
       "      <td>0.010092</td>\n",
       "      <td>0.007008</td>\n",
       "      <td>0.008490</td>\n",
       "      <td>0.008908</td>\n",
       "      <td>0.009444</td>\n",
       "      <td>0.009441</td>\n",
       "      <td>0.009109</td>\n",
       "      <td>0.010956</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005218</td>\n",
       "      <td>0.010238</td>\n",
       "      <td>0.007453</td>\n",
       "      <td>0.019216</td>\n",
       "      <td>0.015510</td>\n",
       "      <td>0.007382</td>\n",
       "      <td>0.011179</td>\n",
       "      <td>0.010922</td>\n",
       "      <td>0.009912</td>\n",
       "      <td>0.007800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>asnC</th>\n",
       "      <td>0.008983</td>\n",
       "      <td>0.009739</td>\n",
       "      <td>0.011383</td>\n",
       "      <td>0.012462</td>\n",
       "      <td>0.009431</td>\n",
       "      <td>0.011511</td>\n",
       "      <td>0.008418</td>\n",
       "      <td>0.013438</td>\n",
       "      <td>0.010727</td>\n",
       "      <td>0.011593</td>\n",
       "      <td>...</td>\n",
       "      <td>0.038385</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.007727</td>\n",
       "      <td>0.011401</td>\n",
       "      <td>0.011182</td>\n",
       "      <td>0.008575</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>0.009574</td>\n",
       "      <td>0.008338</td>\n",
       "      <td>0.008259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>baeR</th>\n",
       "      <td>0.013105</td>\n",
       "      <td>0.008871</td>\n",
       "      <td>0.010121</td>\n",
       "      <td>0.009629</td>\n",
       "      <td>0.009926</td>\n",
       "      <td>0.008589</td>\n",
       "      <td>0.013165</td>\n",
       "      <td>0.009647</td>\n",
       "      <td>0.012880</td>\n",
       "      <td>0.010051</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005879</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.012948</td>\n",
       "      <td>0.012721</td>\n",
       "      <td>0.009071</td>\n",
       "      <td>0.008104</td>\n",
       "      <td>0.012529</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>0.011936</td>\n",
       "      <td>0.008426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>betI</th>\n",
       "      <td>0.013546</td>\n",
       "      <td>0.011041</td>\n",
       "      <td>0.007765</td>\n",
       "      <td>0.015250</td>\n",
       "      <td>0.008617</td>\n",
       "      <td>0.011160</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.010441</td>\n",
       "      <td>0.012913</td>\n",
       "      <td>0.011292</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009014</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>0.033810</td>\n",
       "      <td>0.012868</td>\n",
       "      <td>0.012464</td>\n",
       "      <td>0.006229</td>\n",
       "      <td>0.022091</td>\n",
       "      <td>0.013397</td>\n",
       "      <td>0.009738</td>\n",
       "      <td>0.008143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uxuR</th>\n",
       "      <td>0.010828</td>\n",
       "      <td>0.007333</td>\n",
       "      <td>0.008643</td>\n",
       "      <td>0.015543</td>\n",
       "      <td>0.015179</td>\n",
       "      <td>0.009658</td>\n",
       "      <td>0.013082</td>\n",
       "      <td>0.008641</td>\n",
       "      <td>0.011729</td>\n",
       "      <td>0.008529</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004113</td>\n",
       "      <td>0.006219</td>\n",
       "      <td>0.010012</td>\n",
       "      <td>0.023799</td>\n",
       "      <td>0.009008</td>\n",
       "      <td>0.010766</td>\n",
       "      <td>0.008700</td>\n",
       "      <td>0.011499</td>\n",
       "      <td>0.012824</td>\n",
       "      <td>0.009171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xylR</th>\n",
       "      <td>0.013784</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>0.009421</td>\n",
       "      <td>0.016398</td>\n",
       "      <td>0.013366</td>\n",
       "      <td>0.010478</td>\n",
       "      <td>0.010199</td>\n",
       "      <td>0.016642</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.010524</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046082</td>\n",
       "      <td>0.009321</td>\n",
       "      <td>0.008827</td>\n",
       "      <td>0.016109</td>\n",
       "      <td>0.009885</td>\n",
       "      <td>0.009344</td>\n",
       "      <td>0.011279</td>\n",
       "      <td>0.009494</td>\n",
       "      <td>0.011108</td>\n",
       "      <td>0.016552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydeO</th>\n",
       "      <td>0.008564</td>\n",
       "      <td>0.008821</td>\n",
       "      <td>0.011817</td>\n",
       "      <td>0.010595</td>\n",
       "      <td>0.013121</td>\n",
       "      <td>0.014453</td>\n",
       "      <td>0.009975</td>\n",
       "      <td>0.009100</td>\n",
       "      <td>0.009247</td>\n",
       "      <td>0.013560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013585</td>\n",
       "      <td>0.008773</td>\n",
       "      <td>0.007455</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.008396</td>\n",
       "      <td>0.012985</td>\n",
       "      <td>0.011874</td>\n",
       "      <td>0.013827</td>\n",
       "      <td>0.011802</td>\n",
       "      <td>0.006885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yiaJ</th>\n",
       "      <td>0.008112</td>\n",
       "      <td>0.009439</td>\n",
       "      <td>0.007950</td>\n",
       "      <td>0.009724</td>\n",
       "      <td>0.014005</td>\n",
       "      <td>0.010596</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027658</td>\n",
       "      <td>0.010212</td>\n",
       "      <td>0.013979</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004661</td>\n",
       "      <td>0.011350</td>\n",
       "      <td>0.012163</td>\n",
       "      <td>0.015711</td>\n",
       "      <td>0.011326</td>\n",
       "      <td>0.012519</td>\n",
       "      <td>0.013464</td>\n",
       "      <td>0.010944</td>\n",
       "      <td>0.008091</td>\n",
       "      <td>0.011591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zraR</th>\n",
       "      <td>0.015583</td>\n",
       "      <td>0.008631</td>\n",
       "      <td>0.007820</td>\n",
       "      <td>0.010737</td>\n",
       "      <td>0.008660</td>\n",
       "      <td>0.009412</td>\n",
       "      <td>0.022797</td>\n",
       "      <td>0.023734</td>\n",
       "      <td>0.007028</td>\n",
       "      <td>0.009229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004931</td>\n",
       "      <td>0.007311</td>\n",
       "      <td>0.012432</td>\n",
       "      <td>0.009149</td>\n",
       "      <td>0.008447</td>\n",
       "      <td>0.025233</td>\n",
       "      <td>0.009990</td>\n",
       "      <td>0.007786</td>\n",
       "      <td>0.010017</td>\n",
       "      <td>0.039239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 500 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hipA      hipB      ihfB      yeiL      tdcA      tdcR      yiaJ  \\\n",
       "arcA  0.012169  0.012331  0.010293  0.010130  0.009825  0.015032  0.010608   \n",
       "ascG  0.010982  0.009748  0.010092  0.007008  0.008490  0.008908  0.009444   \n",
       "asnC  0.008983  0.009739  0.011383  0.012462  0.009431  0.011511  0.008418   \n",
       "baeR  0.013105  0.008871  0.010121  0.009629  0.009926  0.008589  0.013165   \n",
       "betI  0.013546  0.011041  0.007765  0.015250  0.008617  0.011160  0.007107   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "uxuR  0.010828  0.007333  0.008643  0.015543  0.015179  0.009658  0.013082   \n",
       "xylR  0.013784  0.012778  0.009421  0.016398  0.013366  0.010478  0.010199   \n",
       "ydeO  0.008564  0.008821  0.011817  0.010595  0.013121  0.014453  0.009975   \n",
       "yiaJ  0.008112  0.009439  0.007950  0.009724  0.014005  0.010596  0.000000   \n",
       "zraR  0.015583  0.008631  0.007820  0.010737  0.008660  0.009412  0.022797   \n",
       "\n",
       "          ompR      bolA      csgD  ...      lldP      rfaY      betB  \\\n",
       "arcA  0.009096  0.005641  0.012541  ...  0.061884  0.009760  0.060858   \n",
       "ascG  0.009441  0.009109  0.010956  ...  0.005218  0.010238  0.007453   \n",
       "asnC  0.013438  0.010727  0.011593  ...  0.038385  0.009619  0.007727   \n",
       "baeR  0.009647  0.012880  0.010051  ...  0.005879  0.011429  0.012948   \n",
       "betI  0.010441  0.012913  0.011292  ...  0.009014  0.008772  0.033810   \n",
       "...        ...       ...       ...  ...       ...       ...       ...   \n",
       "uxuR  0.008641  0.011729  0.008529  ...  0.004113  0.006219  0.010012   \n",
       "xylR  0.016642  0.013736  0.010524  ...  0.046082  0.009321  0.008827   \n",
       "ydeO  0.009100  0.009247  0.013560  ...  0.013585  0.008773  0.007455   \n",
       "yiaJ  0.027658  0.010212  0.013979  ...  0.004661  0.011350  0.012163   \n",
       "zraR  0.023734  0.007028  0.009229  ...  0.004931  0.007311  0.012432   \n",
       "\n",
       "          fliT      rpsS      fucK      rpmC       wzb      nemA      prpD  \n",
       "arcA  0.008531  0.020052  0.005519  0.010090  0.008549  0.011201  0.006739  \n",
       "ascG  0.019216  0.015510  0.007382  0.011179  0.010922  0.009912  0.007800  \n",
       "asnC  0.011401  0.011182  0.008575  0.008481  0.009574  0.008338  0.008259  \n",
       "baeR  0.012721  0.009071  0.008104  0.012529  0.008621  0.011936  0.008426  \n",
       "betI  0.012868  0.012464  0.006229  0.022091  0.013397  0.009738  0.008143  \n",
       "...        ...       ...       ...       ...       ...       ...       ...  \n",
       "uxuR  0.023799  0.009008  0.010766  0.008700  0.011499  0.012824  0.009171  \n",
       "xylR  0.016109  0.009885  0.009344  0.011279  0.009494  0.011108  0.016552  \n",
       "ydeO  0.008390  0.008396  0.012985  0.011874  0.013827  0.011802  0.006885  \n",
       "yiaJ  0.015711  0.011326  0.012519  0.013464  0.010944  0.008091  0.011591  \n",
       "zraR  0.009149  0.008447  0.025233  0.009990  0.007786  0.010017  0.039239  \n",
       "\n",
       "[89 rows x 500 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run GENIE3\n",
    "grns_scores['genie3'] = run_genie3(exp, regulators)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d05812fd-6a4e-43de-b66c-75c2c4dbe8f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on target number 0\n",
      "Working on target number 100\n",
      "Working on target number 200\n",
      "Working on target number 300\n",
      "Working on target number 400\n"
     ]
    }
   ],
   "source": [
    "# Run LASSO\n",
    "grns_scores['lasso'] = run_lasso(exp, regulators)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38d61100-6039-40de-9537-081f41e8ef6a",
   "metadata": {},
   "source": [
    "Next we will convert the matrices with confidence scores to sparse format i.e. from a confidence score matrix to a three column table. \n",
    "\n",
    "| regulator  | target | confidence |\n",
    "|:-:|:-:|:-:|\n",
    "| regulator1 | target_gene1 | score |\n",
    "| regulator3 | target_gene2 | score |\n",
    "| regulator2 | target_gene3 | score |\n",
    "| regulator1 | target_gene1 | score |\n",
    "\n",
    "\n",
    "We will also remove autoregulation (i.e. a regulator which regulates it's own expression) from the networks, remove all interactions with confidence score = 0, and sort the interactions on confidence score.\n",
    "\n",
    "You can use the provided function \"format_network_sparse\" for formatting the networks. You should print out how many interactions each network contains after formatting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "40b3ea37-b37a-45b5-a1bf-0beac2bac85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_network_sparse(confidence_scores):\n",
    "    grn = confidence_scores.reset_index().melt(id_vars=['index'])\n",
    "    grn.columns = ['regulator', 'target', 'confidence']\n",
    "    \n",
    "    # Remove autoregulation\n",
    "    grn = grn[grn.regulator != grn.target]\n",
    "    \n",
    "    # Sort interactions according to confidence\n",
    "    grn = grn.sort_values(by='confidence', ascending=False)\n",
    "    \n",
    "    # Remove interactions with confidence score 0\n",
    "    grn = grn[grn['confidence'] != 0]\n",
    "    return grn\n",
    "\n",
    "# Adjust the code below to format all GRNs to sparse format.\n",
    "grns = {}\n",
    "grns['PCC'] = format_network_sparse(grns_scores['PCC'])\n",
    "grns['genie3'] = format_network_sparse(grns_scores['genie3'])\n",
    "grns['lasso'] = format_network_sparse(grns_scores['lasso'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58cab296-4585-4298-9367-6a20e86d256e",
   "metadata": {},
   "source": [
    "## 4. Selecting the number of interactions in inferred GRNs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e119d0b4-3be4-45ec-bc73-fd0d33497259",
   "metadata": {},
   "source": [
    "The number of interactions in predicted GRNs are often set by either setting a threshold on the number of interactions or setting a threshold on the confidence score. These thresholds are often arbitrarily determined but there are some properties that can be considered when selecting a threshold. We will use two approaches:\n",
    "- Selecting a threshold by comparing the average outdegree for different network cutoffs.\n",
    "- Selecting a threshold by maximizing the pairwise correlation between predicted outdegrees from multiple GRN inference methods.\n",
    "\n",
    "An essential step for using these methods is to calculate the outdegree of all regulators in the GRNs. This can be done using the methods \"group_by\" and \"count\" from Pandas. You can use the provided function \"get_regulator_outdegree\" which generates a pandas DataFrame with the outdegree for each regulator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "90a57a79-09e8-440d-9baf-a00d0b8f1da6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_regulator_outdegree(grn):\n",
    "    regulator_outdegree = grn[['regulator', 'target']].groupby('regulator').count()\n",
    "    return regulator_outdegree\n",
    "\n",
    "regulator_outdegrees = get_regulator_outdegree(grns['PCC'])\n",
    "regulator_outdegrees_genie3 = get_regulator_outdegree(grns['genie3'])\n",
    "regulator_outdegrees_lasso = get_regulator_outdegree(grns['lasso'])\n",
    "np.argmin([np.mean(regulator_outdegrees), np.mean(regulator_outdegrees_genie3), np.mean(regulator_outdegrees_lasso)])\n",
    "\n",
    "## lasso has the lowest avg.out-degree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40f3c79b-3f88-465a-ab1f-2795247f013a",
   "metadata": {},
   "source": [
    "**1. Selecting a threshold by comparing the average outdegree for different network cutoffs**\n",
    "\n",
    "Here you should calculate the average outdegree for each GRN using different cutoffs on the number of interactions in the networks. For example you can start with calculating the average outdegree for the maximum number of interactions and then gradually decrease the threshold, continuously selecting the interactions with the highest confidence score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d35de555-ef8b-46f4-b6bb-f30f704b2a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_interaction = len(grns['lasso'].index)\n",
    "## 2510\n",
    "## average out-degree should be small\n",
    "ninteraction = max_interaction\n",
    "threshold = ninteraction\n",
    "grn_lasso = grns['lasso']\n",
    "grn_lasso = grn_lasso.sort_values(by='confidence', ascending=False).head(ninteraction)\n",
    "mean_regulator_outdegrees = np.mean(get_regulator_outdegree(grn_lasso))\n",
    "step = 1\n",
    "i = 0\n",
    "## yield mean_reg_oudegree\n",
    "## and visualize, maybe find the inflection point\n",
    "\n",
    "def yeild_avg_out_degree(ninteraction, grn_lasso, mean_regulator_outdegrees, step):\n",
    "    while ninteraction > 0:\n",
    "        ninteraction = ninteraction - step\n",
    "        grn_lasso = grn_lasso.sort_values(by='confidence', ascending=False).head(ninteraction)\n",
    "        mean_regulator_outdegree = np.mean(get_regulator_outdegree(grn_lasso))\n",
    "        # i = i + mean_regulator_outdegree < mean_regulator_outdegrees\n",
    "        # if mean_regulator_outdegree < mean_regulator_outdegrees:\n",
    "        #     threshold = ninteraction\n",
    "        mean_regulator_outdegrees = np.min([mean_regulator_outdegree, mean_regulator_outdegrees])\n",
    "        yield mean_regulator_outdegrees\n",
    "mros = yeild_avg_out_degree(ninteraction, grn_lasso, mean_regulator_outdegrees, step)\n",
    "## to list by list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d243bb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "mros_list = list(mros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cd333f3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f1e59614140>]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGdCAYAAACyzRGfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA+SUlEQVR4nO3dd3hUddr/8c9JmxSSkRDSTAgBRFpAaQkQVNyFBQEpFhREsCAooKz604d1fcRdV6w8qwawY0FA3aUJiOJSJBCkhiZNaighhJIGTEhyfn8Es0ZqIJkzM3m/rutclzlzkrnzdS7n48n3vscwTdMUAACAk3hZXQAAAKheCB8AAMCpCB8AAMCpCB8AAMCpCB8AAMCpCB8AAMCpCB8AAMCpCB8AAMCpfKwu4PdKSkp08OBBBQcHyzAMq8sBAACXwTRN5eXlKTo6Wl5eF7+34XLh4+DBg4qNjbW6DAAAcAUyMjIUExNz0WtcLnwEBwdLKi0+JCTE4moAAMDlyM3NVWxsbNn7+MW4XPj49U8tISEhhA8AANzM5WyZYMMpAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwKsIHAABwqmoVPl75dqsmLP5FJSWm1aUAAFBtudyn2laVDftP6N0lOyVJa/ce15t33SB7oK/FVQEAUP1UmzsfCdfaNbZvgvx8vPTDliz1SFmqTQdyrC4LAIBqp9qED8MwdG/bOpr+aHvFhgYo49gp9Z24XFNX7pNp8mcYAACcpULhY+LEiWrevLlCQkIUEhKidu3a6dtvvy173DRNjRkzRtHR0QoICNAtt9yizZs3V3rRV6PZtXbNGdFRf2wcrsKiEo2evlFPf71BpwqLrS4NAIBqoULhIyYmRq+88opWr16t1atX69Zbb1WvXr3KAsZrr72mcePGKSUlRatWrVJkZKQ6d+6svLy8Kin+StkDffX+wNZ6puv18jKkf6/drz4Tlml3doHVpQEA4PEM8yr/5hAaGqrXX39dDz74oKKjozVq1Cg9++yzkiSHw6GIiAi9+uqrGjp06GX9vNzcXNntduXk5CgkJORqSrssy3dm6/Gp65SdX6gaNh+9cVdzdW0WVeXPCwCAJ6nI+/cV7/koLi7WtGnTVFBQoHbt2mn37t3KzMxUly5dyq6x2Wy6+eabtXz58gv+HIfDodzc3HKHM7WvH6a5j3dU27qhyncUadjktXppzs86U1zi1DoAAKguKhw+Nm7cqBo1ashms2nYsGGaMWOGmjRposzMTElSREREuesjIiLKHjufsWPHym63lx2xsbEVLemqRYT464shiXrkpnqSpA9Td6v/Byt0OPe002sBAMDTVTh8XH/99UpPT9eKFSv06KOPatCgQfr555/LHjcMo9z1pmmec+63Ro8erZycnLIjIyOjoiVVCl9vL/3ltsZ6975WCrb5aNWe4+r+9lIt35ltST0AAHiqCocPPz8/NWjQQK1bt9bYsWPVokULvfXWW4qMjJSkc+5yZGVlnXM35LdsNltZ98yvh5W6NovU7JHJahQZrOz8Qt334U8av4ipqAAAVJarnvNhmqYcDofi4+MVGRmpBQsWlD1WWFioJUuWqH379lf7NE4VHxakmcM76K5WMSoxpde/26Yhn61WzskzVpcGAIDbq1D4+Mtf/qKlS5dqz5492rhxo5577jktXrxYAwYMkGEYGjVqlF5++WXNmDFDmzZt0uDBgxUYGKj+/ftXVf1Vxt/XW6/f1UKv3lE6FfU/W7PU/Z2l2rifqagAAFyNCn22y+HDhzVw4EAdOnRIdrtdzZs31/z589W5c2dJ0jPPPKNTp07pscce0/Hjx5WYmKjvv/9ewcHBVVK8M/RrU0dNo+167Iu12nfspO6YuFxjbm+qe9vGXnQvCwAAOL+rnvNR2Zw95+Ny5Zw6o6e+Wq8fthyWJPW98Vr9o0+CAvy8La4MAADrOWXOR3VjD/DVB/e30v90ayQvQ5q+7oB6j1+mXUfyrS4NAAC3QvioAMMwNOzm+poyJElhNWzadjhPt6cs07yNh6wuDQAAt0H4uAJJ9Wpp3uPJahtfOhX1sS/W6u9MRQUA4LIQPq5QeIi/pjycqKFnp6J+lLpb97y/Qpk5TEUFAOBiCB9XwcfbS6Nva6z3BrZSsL+P1uwtnYq67BemogIAcCGEj0rwp6aRmjMyWY2jQnS0oFADP/pJKQt3MBUVAIDzIHxUkrhaQZrxWHvd3bp0Kuob32/XQ5+u0omThVaXBgCASyF8VCJ/X2+9dmcLvXZHc9l8vLRo2xF1fztVG/afsLo0AABcBuGjCtzdJlbTH2uvuFqBOnDilO6cmKbJK/bKxea5AQBgCcJHFWkabdfsEcnq3CRChcUl+uvMTXryq/U6WVhkdWkAAFiK8FGF7AG+en9g6VRUby9DM85ORd3JVFQAQDVG+Khiv05F/eLhRNUOtmn74Xzd/k6q5m5gKioAoHoifDhJUr1amjuydCpqQWGxhk9Zqxe/2azCIqaiAgCqF8KHE5VNRb25dCrqpGV7dM/7aTqUc8riygAAcB7Ch5P5eHtpdLfGev/sVNS1+06o+9upSt3BVFQAQPVA+LBIl7NTUZtEhehYQaEGfvyT3v4PU1EBAJ6P8GGhuFpBmv5Ye/VrHSvTlMYt2K4HP12l4wVMRQUAeC7Ch8X8fb316p3N9dqdpVNRF287oh7vpGp9xgmrSwMAoEoQPlzE3a3LT0W96900fc5UVACAByJ8uJCm0XZ9MzJZXc5ORX1+5ib9+ct0pqICADwK4cPFhPj76r2BrfSX20qnos5MP6heKcv0SxZTUQEAnoHw4YIMw9AjN9XXlLNTUXdk5atXSqrmbDhodWkAAFw1wocLS6xXS3MfT1ZSvdKpqCOmrNOY2UxFBQC4N8KHiwsP9tfkhxL16C31JUmfLN+jfu+n6eAJpqICANwT4cMN+Hh76dmujfTB/a0V7O+jdftOqMc7qVq644jVpQEAUGGEDzfSuUmE5o7sqKbRpVNR7/94pd76gamoAAD3QvhwM3VqBerfj7bXvW1Lp6L+3w/bNfiTVTrGVFQAgJsgfLghf19vje3bXG/c1UI2Hy/9uP2Ier6TqnSmogIA3ADhw43d2SpGM4d3UN2yqajL9XnaHqaiAgBcGuHDzTWOCtHskcnq2jRSZ4pNPT9rs0Z9ma4CB1NRAQCuifDhAUL8fTXxvpb6a/fG8vYyNCv9oHqPX6ZfsvKsLg0AgHMQPjyEYRh6uGM9TXskSeFnp6LenrJM36xnKioAwLUQPjxMm7qhmvt4R7WrV0snC4s1cipTUQEAroXw4YFqB9v0+UNt9dhvpqLe/R5TUQEAroHw4aF8vL30TNdG+mhQa4X4+yg944S6v71UP25nKioAwFqEDw/3h8YRmvt4RzW7NkTHT57RoEkr9c8ftjMVFQBgGcJHNRAbGqh/DWuve9vWkWlK//xhB1NRAQCWIXxUE6VTURP05l0t5O9bOhW1x9tLtW7fcatLAwBUM4SPauaOs1NR48OCdDDntO5+L02fLmcqKgDAeQgf1VCjyBDNHtFB3ZqVTkV9YfZmPT6NqagAAOcgfFRTwf6+mjCgdCqqj5ehb9YfVC+mogIAnIDwUY39dipqRIhNv5ydijor/YDVpQEAPBjhA2p9dipq+/qlU1GfmJau/521SY6iYqtLAwB4IMIHJElhNWz6/KFEjejUQJL0Wdpe3f3eCh1gKioAoJIRPlDG28vQ03+6Xh8Pbi17gK/WZ5xQj7eXaglTUQEAlYjwgXPc2ihCc0YmK+Fau46fPKPBk1Zq3ILtKmYqKgCgElQofIwdO1Zt2rRRcHCwwsPD1bt3b23btq3cNYMHD5ZhGOWOpKSkSi0aVS82NFBfD2unAYmlU1Hf/s8ODZ60kqmoAICrVqHwsWTJEg0fPlwrVqzQggULVFRUpC5duqigoKDcdV27dtWhQ4fKjnnz5lVq0XAOf19v/aNPgsbdXToVdemObHV/e6nWMhUVAHAVfCpy8fz588t9PWnSJIWHh2vNmjW66aabys7bbDZFRkZWToWwXN+WMWoabdejk9doV3aB+r2Xpudua6xB7evKMAyrywMAuJmr2vORk5MjSQoNDS13fvHixQoPD1fDhg01ZMgQZWVlXfBnOBwO5ebmljvgeq6PDNasER10W0LpVNQx3/yskVPXKZ+pqACACjLMK/xQD9M01atXLx0/flxLly4tO//ll1+qRo0aiouL0+7du/X888+rqKhIa9askc1mO+fnjBkzRi+++OI553NychQSEnIlpaEKmaapj5ft0dh5W1RUYqp+7SBNvK+VGkYEW10aAMBCubm5stvtl/X+fcXhY/jw4Zo7d65SU1MVExNzwesOHTqkuLg4TZs2TX379j3ncYfDIYfDUa742NhYwoeLW73nmIZPWavDuQ4F+HrrlTsS1OuGa60uCwBgkYqEjyv6s8vIkSM1e/ZsLVq06KLBQ5KioqIUFxenHTt2nPdxm82mkJCQcgdc369TUTs0qKVTZ0qnoj4/k6moAIBLq1D4ME1TI0aM0PTp07Vw4ULFx8df8nuOHj2qjIwMRUVFXXGRcE1hNWz67MFEjby1dCrq5yv26u5307T/+EmLKwMAuLIKhY/hw4dr8uTJmjJlioKDg5WZmanMzEydOlU6gjs/P19PP/200tLStGfPHi1evFg9e/ZUWFiY+vTpUyW/AKzl7WXoqS7Xa9LgNqVTUffnqMc7qVq07cKbjAEA1VuF9nxcqK1y0qRJGjx4sE6dOqXevXtr3bp1OnHihKKiotSpUyf9/e9/V2xs7GU9R0X+ZgTXknHspIZPWasN+3NkGNLITg30xB8bytuLdlwA8HRO2XBaVQgf7s1RVKy/ffOzvvhpnySp43Vh+me/G1SrxrmdTgAAz1HlG06BC7H5lE5F/b9+LRTg6312Kmqq1uxlKioAoBThA1Wiz40xmjm8g+rVDlJm7mn1ey9NH6fulovdaAMAWIDwgSpzfWSwZo9IVveEKBWVmPrbnJ81gqmoAFDtET5QpWrYfJTS/0b9b48m8vEyNHfDId2ekqrth/OsLg0AYBHCB6qcYRh6MDleXw5tp8gQf+06UqBeKcs0Y91+q0sDAFiA8AGnaRVXU3MfT1ZygzCdOlOsP3+5Xs/N2MhUVACoZggfcKpaNWz69MG2evzsVNQvftqnu95NU8YxpqICQHVB+IDTeXsZerLL9Zr0QBtdE+irDb9ORd3KVFQAqA4IH7BMp+vDNWdkslrE2JVz6owe+GSV3vhum4pLaMcFAE9G+IClYmoG6qth7TQwKU6SlLLoF93/8U/KzndYXBkAoKoQPmA5m4+3/t67md665wYF+Hpr2S9H1ePtVK3Ze8zq0gAAVYDwAZfR64ZrNWvEb6eirtBHTEUFAI9D+IBLaRhROhW1R/PSqah/n/Ozhk9Zq7zTZ6wuDQBQSQgfcDk1bD56594bNaZnE/l6G5q3MVO9UpZpa2au1aUBACoB4QMuyTAMDe5QOhU1yu6vXdkF6j1+maavZSoqALg7w3SxP6jn5ubKbrcrJydHISEhVpcDF3A036FRX6Zr6Y5sSVJYDZu8jNLHfLwM9brxWj3ZuaF8vcnSAGCVirx/819ruLxaNWz65IG2euIP18kwpOx8h7LySo+DOac1cfFODfjgJ2Xlnba6VADAZeDOB9xKVu5pZecXln29/XCenp+5SXmOItUOtmnCgJZqUzfUwgoBoHqqyPs34QNub9eRfA2bvEbbD+fLx8vQ6Nsa68EOdWUYhtWlAUC1wZ9dUK3Uq11DM4d30O0tosvac0dOXacCR5HVpQEAzoPwAY8Q6Oejt+65QWN6NpGPl6E5Gw6p1/hl+iUr3+rSAAC/Q/iAx/hve26SIkJs+iUrX71SUvXtxkNWlwYA+A3CBzxOq7hQzRnZUYnxoSooLNajX6zVy/O2qKi4xOrSAAAifMBD1Q626YuHEzX0pnqSpPd/3KUBH9KOCwCugPABj+Xj7aXRtzXWxAEtVcPmo592H1OPt1O1eg+flgsAViJ8wON1S4jSrBEddF14DWXlOXTP+ys0aRmflgsAViF8oFqof7Yd99dPy33xm5/1xLR02nEBwAKED1QbQWc/Lfd/e5S2485ef1B9JizTriO04wKAMxE+UK0YhqEHk+M19ZEkhQfbtP1wvm5PWab5m2jHBQBnIXygWmpTN1RzHk9W2/hQ5TuKNGzyWo39lnZcAHAGwgeqrfBgf33xcKKGdIyXJL23ZJcGfrRSR/IcFlcGAJ6N8IFqzdfbS891b6IJA1oqyM9babuOqsc7S7Vm73GrSwMAj0X4ACTddrYdt0F4DR3Odeie99P06fI9tOMCQBUgfABnNQgP1szhHdQ9IUpnik29MHuzRn2ZrpOFtOMCQGUifAC/UcPmo5T+N+qv3RvL28vQrPSD6jN+uXZnF1hdGgB4DMIH8DuGYejhjvU0dUiSagfbtO1wnm5/J1Xfbc60ujQA8AiED+AC2saHau7IZLWpW1N5jiIN/XyNXvl2K+24AHCVCB/ARYSH+GvKkCQ9lFzajvvukp26/+OVys6nHRcArhThA7gEX28vPd+jiVL636hAP28t33lUPd5O1dp9tOMCwJUgfACXqUfzaM0a3kH1agcpM/e0+r2Xps/TaMcFgIoifAAVcF1EsGYN76BuzSJ1ptjU87M268mv1utUYbHVpQGA2yB8ABUU7O+rCQNa6rnbSttxZ6w7oD4TlmkP7bgAcFkIH8AVMAxDQ26qpy8eTlRYDZu2ZuapZ0qqFvx82OrSAMDlET6Aq5BUr5bmPp6sVnE1lXe6SEM+W63Xv9uq4hL2gQDAhRA+gKsUEeKvqUOSNLh9XUnS+EU7NejjlTpKOy4AnBfhA6gEfj5eGnN7U711zw0K8PVW6i/Z6vlOqtIzTlhdGgC4nAqFj7Fjx6pNmzYKDg5WeHi4evfurW3btpW7xjRNjRkzRtHR0QoICNAtt9yizZs3V2rRgKvqdcO1mjWig+qFBelgzmnd9e5yTV6xl3ZcAPiNCoWPJUuWaPjw4VqxYoUWLFigoqIidenSRQUF/93l/9prr2ncuHFKSUnRqlWrFBkZqc6dOysvL6/SiwdcUcOIYM0a0UF/ahqhM8Wm/jpzk576mnZcAPiVYV7F/5IdOXJE4eHhWrJkiW666SaZpqno6GiNGjVKzz77rCTJ4XAoIiJCr776qoYOHXrJn5mbmyu73a6cnByFhIRcaWmA5UzT1Ps/7tKr87eqxJQaR4Xo3ftaKq5WkNWlAUClq8j791Xt+cjJyZEkhYaGSpJ2796tzMxMdenSpewam82mm2++WcuXLz/vz3A4HMrNzS13AJ7AMAwNvbm+Jj+cqLAaftpyKFc93knVf7bQjgugervi8GGapp588kklJyerWbNmkqTMzNKPHI+IiCh3bURERNljvzd27FjZ7fayIzY29kpLAlxS+/phmjOyo1rWuUZ5p4v00Ker9cZ322jHBVBtXXH4GDFihDZs2KCpU6ee85hhGOW+Nk3znHO/Gj16tHJycsqOjIyMKy0JcFmRdn9Ne6RdWTtuyqJfNHjSSh0rKLS2MACwwBWFj5EjR2r27NlatGiRYmJiys5HRkZK0jl3ObKyss65G/Irm82mkJCQcgfgiX5tx/1nv9J23KU7Sttx19OOC6CaqVD4ME1TI0aM0PTp07Vw4ULFx8eXezw+Pl6RkZFasGBB2bnCwkItWbJE7du3r5yKATfX+8ZrNWN4e9WtFagDJ07prnfTNOWnfbTjAqg2KhQ+hg8frsmTJ2vKlCkKDg5WZmamMjMzderUKUmlf24ZNWqUXn75Zc2YMUObNm3S4MGDFRgYqP79+1fJLwC4o0aRIZo9MlldmkSosLhEf5mxUf/vXxt0+gztuAA8X4VabS+0b2PSpEkaPHiwpNK7Iy+++KLee+89HT9+XImJiRo/fnzZptRLodUW1Ylpmnp3yS69/l1pO26TqBC9e18r1akVaHVpAFAhFXn/vqo5H1WB8IHqaPkv2Ro5dZ2OFhQqxN9Hb91zozo1Cre6LAC4bE6b8wGgcrRvEKY5jyfrhthrlHu6SA98skrjFmynHReARyJ8AC4iyh6gL4cmaWBSnCTp7f/s0AOfrNJx2nEBeBjCB+BCbD7e+nvvZvq/fi3k7+ulH7cfUY93UrVh/wmrSwOASkP4AFxQnxtjNOOxDoo7245758Q0TVu5z+qyAKBSED4AF9U4KkSzRyTrj41L23H/Z/pGPfOv9bTjAnB7hA/AhdkDfPX+wFb6f3+6Xl6G9NXq/brz3eXKOHbS6tIA4IoRPgAX5+VlaHinBvrswUSFBvlp04HST8ddtC3L6tIA4IoQPgA3kXxdmOaMTFaL2GuUc+qMHvxklf75w3aV0I4LwM0QPgA3En1NgL4amqQBiXVkmtI/f9ihBz9dpRMnaccF4D4IH4Cbsfl46x99EvTmXS1k8/HS4m2l7bibDuRYXRoAXBbCB+Cm7mhV2o5bJzRQ+4+fUt+Jy/XVqgyrywKASyJ8AG6sSXSIvhmRrD80CldhUYme+fcGjZ7Op+MCcG2ED8DN2QN99cH9rfVU54YyDGnqygzd9W6a9h+nHReAayJ8AB7Ay8vQyD9cp08faKuagb7aeCBHPd5J1ZLtR6wuDQDOQfgAPMhNDWvrm5HJah5j14mTZzR40kq9/Z8dtOMCcCmED8DDxNQM1FdD2+netqXtuOMWbNfDn61WzskzVpcGAJIIH4BH8vf11ti+CXrtzuay+Xhp4dYs9UhZSjsuAJdA+AA82N2tY/XvR9srNjRAGcdO6Y6Jy/X1atpxAViL8AF4uGbX2jVnREd1ur62HEUl+n//2qDR0zfKUUQ7LgBrED6AasAe6KuPBrXRk2XtuPt097tpOnDilNWlAaiGCB9ANeHlZejxP1ynTx5oq2sCfbV+f456vL1US3fQjgvAuQgfQDVzc8Pa+mZEshKutev4yTO6/+OVSllIOy4A5yF8ANVQbGigvh7WTve0iZVpSm98v12PfL5aOadoxwVQ9QgfQDXl7+utV+5ortfuaC4/Hy/9sCVLPd9J1c8Hc60uDYCHI3wA1dzdbWI1/dH2iqkZoH3HTqrPhGX695r9VpcFwIMRPgCUtuOOTNbNDUvbcZ/6er3+OpN2XABVg/ABQJJ0TaCfJg1uoyf+cJ0MQ5q8Yp/ufm+FDtKOC6CSET4AlPHyMvTnzg318eA2sgf4an3GCfV4J1WpO7KtLg2AByF8ADhHp+vDNWdksppGh+hYQaHu//gnjV/0C+24ACoF4QPAecWGBurfj7bX3a1jVGJKr3+3TY98voZ2XABXjfAB4IL8fb312p0t9ErfhLPtuIfVKyVVWw7RjgvgyhE+AFzSPW3r6F/D2unaawK052hpO+6MdbTjArgyhA8Al6V5zDWaMzJZNzWsrdNnSvTnL9fr+ZmbVFhUYnVpANwM4QPAZasZVNqO+/itDSRJn6/Yq37vp+lQDu24AC4f4QNAhXh7GXqyy/X6eHBrhfj7aN2+E+rxdqqW/0I7LoDLQ/gAcEVubRShOSM7qklUiI4WFOq+j37SxMU7ZZq04wK4OMIHgCtWp1agpj/WXne0LG3HfXX+Vg39fI1yT9OOC+DCCB8Aroq/r7feuKu5/tGnmfy8vfT9z4fVK2WZtmXmWV0aABdF+ABw1QzD0IDEOH09rJ2i7f7anV2g3uOXaVb6AatLA+CCCB8AKk2L2Gs05/GO6nhdmE6dKdYT09I1ZvZm2nEBlEP4AFCpQoP89MkDbTWiU2k77ifL9+ie99OUmXPa4soAuArCB4BK5+1l6Ok/Xa8P72+tYH8frd13Qj3eWarlO2nHBUD4AFCF/tgkQt+MSFajyGBl5xfqvg9/0ntLaMcFqjvCB4AqVTcsSDMe66C+La9ViSmN/XarHp28Vnm04wLVFuEDQJUL8PPWm3e10Eu9m8nX29D8zZnqlbJM2w/TjgtURxUOHz/++KN69uyp6OhoGYahmTNnlnt88ODBMgyj3JGUlFRZ9QJwU4Zh6L6kOH01tJ2i7P7alV2gXinLNHv9QatLA+BkFQ4fBQUFatGihVJSUi54TdeuXXXo0KGyY968eVdVJADPcWOdmpozMlkdGtTSqTPFenzqOr34zWadKaYdF6gufCr6Dd26dVO3bt0ueo3NZlNkZOQVFwXAs9WqYdNnDybqze+3acLinZq0bI827s/R+AEtFRHib3V5AKpYlez5WLx4scLDw9WwYUMNGTJEWVlZVfE0ANyYt5ehZ7o20vsDWynY5qPVe4+r+9upWrHrqNWlAahilR4+unXrpi+++EILFy7Um2++qVWrVunWW2+Vw+E47/UOh0O5ubnlDgDVR5emkZo98td2XIcGfPiTPvhxF+24gAer9PDRr18/de/eXc2aNVPPnj317bffavv27Zo7d+55rx87dqzsdnvZERsbW9klAXBx8WFBmv5Ye/W58VoVl5j6x7wtGj5lrfIdRVaXBqAKVHmrbVRUlOLi4rRjx47zPj569Gjl5OSUHRkZGVVdEgAXFOjno3F3t9DfejWVr7eheRsz1SslVb9k0Y4LeJoqDx9Hjx5VRkaGoqKizvu4zWZTSEhIuQNA9WQYhu5vV1dfDm2nyBB/7TxSoNtTlmnOBtpxAU9S4fCRn5+v9PR0paenS5J2796t9PR07du3T/n5+Xr66aeVlpamPXv2aPHixerZs6fCwsLUp0+fyq4dgIdqWaem5jyerHb1aulkYbFGTFmnv33zM+24gIcwzAru6lq8eLE6dep0zvlBgwZp4sSJ6t27t9atW6cTJ04oKipKnTp10t///vfL3suRm5sru92unJwc7oIA1VxRcYne+H673l2yU5LUpm5Nje/fUuG04wIupyLv3xUOH1WN8AHg977bnKmnv1qvPEeRagfbNL5/S7WND7W6LAC/UZH3bz7bBYDL+1PTSM0a0UHXRwTrSJ5D936wQh8upR0XcFeEDwBuoV7tGpoxvL163RCt4hJTL83dohFT19GOC7ghwgcAtxHo56N/9rtBY3o2kY+XobkbDqn3+GW04wJuhvABwK0YhqHBHeL15dAkRYTY9EtWvnqlLNO8jYesLg3AZSJ8AHBLreJCNWdkRyXVC1VBYbEe+2Kt/jH3ZxXRjgu4PMIHALdVO9imyQ8lauhN9SRJHyzdrf4f/qSsvNMWVwbgYggfANyaj7eXRt/WWBMHtFQNm49W7j6mHm+natWeY1aXBuACCB8APEK3hCjNGtFB14XXUFaeQ/e+v0Ifp+6mHRdwQYQPAB6jfu0amjm8g3q2iFZRiam/zflZj09LVwHtuIBLIXwA8ChBNh+9fc8NeuFsO+436w+q9/hl2nkk3+rSAJxF+ADgcQzD0AMd4jXtkSSFB9u042w77re04wIugfABwGO1rhuqOY8nq218qPIdRXr0i7V6ed4W2nEBixE+AHi08GB/ffFwooZ0jJckvf/jLt330U86kuewuDKg+iJ8APB4vt5eeq57E00Y0FJBft5aseuYeryzVGv20o4LWIHwAaDauC0hSrNGJKtBeA0dznWo33sr9Mky2nEBZyN8AKhWGoSXtuN2T4hSUYmpMd/8rCempetkIe24gLMQPgBUOzVsPkrpf6P+2r2xvL0MzV5/UH3GL9cu2nEBpyB8AKiWDMPQwx3raeqQJNUOtmnb4Tz1Slmm+ZsyrS4N8HiEDwDVWtv4UM0dmaw2dWsqz1GkYZPX6JVvt9KOC1QhwgeAai88xF9ThiTpoeTSdtx3l+zUwI9WKjufdlygKhA+AECl7bjP92iilP43KtDPW2m7jqrH26lau++41aUBHofwAQC/0aN5tGYN76D6tYOUmXta/d5L02dpe2jHBSoR4QMAfue6iGDNGpGs2xIidabY1P/O2qwnv1pPOy5QSQgfAHAeNWw+Gt+/pZ67rbQdd8a6A+o7Ybl2ZxdYXRrg9ggfAHABhmFoyE319MXDiQqrYdPWzDzd/k6qvt9MOy5wNQgfAHAJSfVqae7jyWodV9qO+8jna/Ta/K0qLmEfCHAlDNPFdlHl5ubKbrcrJydHISEhVpcDAGXOFJfo5XlbNGnZHklSeLBNAX7euiH2Gr14e1NdE+hnbYGAhSry/s2dDwC4TL7eXnqhZ1O9dc8NCvD1VlaeQ3uPntSs9IPqmZKqzQdzrC4RcAvc+QCAK3CsoFC7swuUe/qMXpi1WfuOnZTNx0tj+yaob8sYq8sDnI47HwBQxUKD/NQqrqY6XR+ub0Yk65bra8tRVKInv1qvF2ZtUmER49mBCyF8AMBVsgf66uNBbfT4H66TJH2atlf9P1ihrNzTFlcGuCbCBwBUAi8vQ092bqgP72+tYJuPVu89ru7vpGr1nmNWlwa4HPZ8AEAl251doKGfr9b2w/ny9jIUGeKvQD9v9WsTq4eS42UYhtUlApWOPR8AYKH4sCDNeKyDejSPUnGJqQMnTmlHVr5emrtFw6esVYGDMe2o3rjzAQBVaE92gU6cOqO1e49r7LdbdKbY1PURwXpvYCvVDQuyujyg0nDnAwBcRN2wIN0Qe40eTI7XtEeSVDvYpm2H83R7SqoWb8uyujzAEoQPAHCSVnGhmjMyWS3rXKPc00V64JNVGr/oF7nYDWigyhE+AMCJIkL8NfWRJN3bto5MU3r9u2167Av2gaB6IXwAgJPZfLw1tm+CXu6TIF9vQ99uylSfCcu0J7vA6tIApyB8AIBF+ifW0bRH2ik82Kbth/N1e0qqFrEPBNUA4QMALNQqrqbmjExWq7iayj1dpAfZB4JqgPABABYLD/HX1CFJ6p/4330gj05eq3z2gcBDET4AwAX4+Xjp5T4JGts3QX7eXpq/OVN9xi/TbvaBwAMRPgDAhdzbto6mDU1SeLBNO7LO7gPZyj4QeBbCBwC4mJZ1/rsPJO90kR78dJVSFu5gHwg8BuEDAFzQr/tABpzdB/LG99vZBwKPUeHw8eOPP6pnz56Kjo6WYRiaOXNmucdN09SYMWMUHR2tgIAA3XLLLdq8eXNl1QsA1Yafj5f+0SdBr7APBB6mwuGjoKBALVq0UEpKynkff+211zRu3DilpKRo1apVioyMVOfOnZWXl3fVxQJAdXTP2X0gESH/3QeycOthq8sCrthVfaqtYRiaMWOGevfuLan0rkd0dLRGjRqlZ599VpLkcDgUERGhV199VUOHDr3kz+RTbQHg/LLyTuuxyWu1eu9xGYY0/JYGah5jV2xooBpH8d9LWMuyT7XdvXu3MjMz1aVLl7JzNptNN998s5YvX37e73E4HMrNzS13AADOFR7srylDknRfUuk+kJRFv+iRz9eo21tL9fp3W1VcwoZUuIdKDR+ZmZmSpIiIiHLnIyIiyh77vbFjx8put5cdsbGxlVkSAHgUPx8vvdQ7QW/e1UJt40PVPMYuSRq/aKeGfLZauafPWFwhcGlV0u1iGEa5r03TPOfcr0aPHq2cnJyyIyMjoypKAgCPckerGH01tJ1mj0jWP/vdIJuPlxZuzVLvlGX6JSvf6vKAi6rU8BEZGSlJ59zlyMrKOuduyK9sNptCQkLKHQCAy9f7xmv1r2HtFW33167sAvUZv0z/2cKGVLiuSg0f8fHxioyM1IIFC8rOFRYWasmSJWrfvn1lPhUA4DcSYuyaPTJZbeuGKs9RpIc/W81gMrisCoeP/Px8paenKz09XVLpJtP09HTt27dPhmFo1KhRevnllzVjxgxt2rRJgwcPVmBgoPr371/ZtQMAfiOshk2TH04s25D6xvfbNXzKWhUwmAwupsKttosXL1anTp3OOT9o0CB98sknMk1TL774ot577z0dP35ciYmJGj9+vJo1a3ZZP59WWwC4elNX7tP/ztqkM8WmGkUG64P7Wys2NNDqsuDBKvL+fVVzPqoC4QMAKsfqPcc0bPJaZec7VDPQV+P7t1T7BmFWlwUPZdmcDwCA62hdN1TfjOyg5jF2HT95RgM/XqmPU3ezDwSWI3wAgAeLsgfoq6Ht1PfGa1VcYupvc37W019v0OkzxVaXhmqM8AEAHs7f11tv3t1Cf+3eWF6G9O+1+9Xv/RXKzDltdWmopggfAFANGIahhzvW06cPtpU9wFfrM06oZ0qq1uw9bnVpqIbYcAoA1czeowV65LM12nY4T37eXrovKU61avip1w3RiqlJRwyuDN0uAICLKnAU6amv1mv+5v9OpL4m0FcT6IjBFSJ8AAAuqaTE1NdrMrT5YK7W7D2uzQdz5e1laOhN9RQbGqhWcTUVfU2Aath8rC4VboDwAQCokNNnijV6+kbNWHfgnMe6NYvUG3e1UBAhBBdRkfdvXkkAAPn7emvc3S2UVC9US3dka8fhfG07nCdJ+nZTpvYcPakPB7XWtdcEWFwpPAF3PgAA52WaptZlnNAjn61Rdr5DYTVs+uD+VrqxTk2rS4MLYsIpAOCqGYahlnVqataIDmoUGazsfIfueX+Fvll/0OrS4OYIHwCAi7r2mgD969H2+mPjcDmKSjRy6jq99cMOxrTjihE+AACXVMPmo/cGttaQjvGSpP/7YbuemJbOmHZcEcIHAOCyeHsZeq57E73SN0E+XoZmrz+oez9YoSN5DqtLg5shfAAAKuSetnX02UOlY9rX7Tuh3uOXaWtmrtVlwY0QPgAAFda+fphmDu+gemFBOnDilO6YsFwLtx62uiy4CcIHAOCKxIcFacZjHdS+fi0VFBbr4U9X68Olu9iIiksifAAArpg90FefPthW97atoxJTemnuFv1lxkadKS6xujS4MMIHAOCq+Hp76eU+zfTX7o1lGNLUlRka9PFKnThZaHVpcFGEDwDAVTMMQw93rKcP72+tID9vLd95VH0mLNeavcd0NJ9uGJTHeHUAQKXacihXD3+6WgdOnJIkGYY09Kb6euZP18vLy7C4OlQVxqsDACzTOCpEM4d30K2NwhUa5CfTlN5dslPDp6zVqUKGkoE7HwCAKjZj3X49+6+NKiwuUYsYuz4Y1Frhwf5Wl4VKxp0PAIDL6HNjjCY/nKiagb5avz9HfcYvZyhZNUf4AABUubbxoZrx2H+Hkt05MU2Lt2VZXRYsQvgAADhF3bAgTX+svZLqhSrfUaQHP1mlz9P2WF0WLED4AAA4zTWBfvrswUTd2SpGJab0/KzNenzqOr27ZCdzQaoRNpwCAJzONE1NWLxTr3+3rexclN1fjaNC9FByvDo0CLOwOlyJirx/+zipJgAAyhiGoeGdGqh5jF2Ltx3R/E2ZOnDilA7lnNaS7Uf0cp9m6temjtVloopw5wMAYLmcU2e0/Jdszd+cqVnpByVJj9/aQH/u3FCGwWAyd8CdDwCAW7EH+KpbQpS6NotUXGig3l74i95e+Iv2nzilgUlxahptl58P2xQ9BXc+AAAuZ9rKfXpu5iYVl5S+RdWrHaSxfRLUKCpE9gBfi6vD+VTk/ZvwAQBwSUu2H9H/LdiunUfylXe6SJIU4Outd+69UX9sEmFxdfg9wgcAwGNk5pzWX2Zs1Ib9OcrOd8jby9DYPgm6u02s1aXhNxivDgDwGJF2f308uI3SRt+qO1vFqLjE1DP/3qCUhTvkYv//jMtE+AAAuAVfby+9fmdzPXZLfUnSG99v1wuzN5ftC4H7IHwAANyGYRh6pmsjjenZRIYhfZa2VyOnrtXpM8VWl4YKIHwAANzO4A7xeufeG+XrbWjexkwNnrRSuafPWF0WLhPhAwDglno0j9anD7RVDZuPVuw6pn7vrVBW7mmry8JlIHwAANxW+wZhmvZIksJq2LTlUK76TlyuXUfyrS4Ll0CrLQDA7e07elL3f/yT9hw9qdAgP43o1KBKJqIG+/voT00j5e/rXek/290x5wMAUO1k5zv0wKRV2nggp0qfJ6leqD5/KFG+3vzx4LcIHwCAaqnAUaS3/rND+46erPSfXVhcooVbs8q+/mv3xnq4Y71Kfx53xQfLAQCqpSCbj/5yW+Mq+/nzNx3SqC/TdfpMif4xb4saRgTrpoa1q+z5PBX3jAAAuExdm0Vp45g/6Y6WMTJN6c9fptNhcwUIHwAAVICvt5f+0aeZGkUG62hBoZ6Yls6U1Qqq9PAxZswYGYZR7oiMjKzspwEAwDL+vt4aP6ClAv28lbbrqN5ZuMPqktxKldz5aNq0qQ4dOlR2bNy4sSqeBgAAy9SvXUP/6NNMkvTWf3Zo+c5siytyH1USPnx8fBQZGVl21K7NZhwAgOfpc2OM7m5duv/jiWnpys53WF2SW6iSbpcdO3YoOjpaNptNiYmJevnll1Wv3vnbkRwOhxyO//7Lys3NrYqSAACoEmNub6p1+05oR1a+El/+j7wNo9zjNfx91L9tHcXUDJCPt5d8vAzVr11DCTF2iyq2XqXP+fj222918uRJNWzYUIcPH9ZLL72krVu3avPmzapVq9Y5148ZM0YvvvjiOeeZ8wEAcBfbD+dpwIc/6Uje5d358DKkKUOS1Oxau2rYPGPqhUsNGSsoKFD9+vX1zDPP6Mknnzzn8fPd+YiNjSV8AADcyukzxTp+srDcuaxchz5dvkcFhUUqLjFVVGJq39GT2pVdIEkyDOmFHk00uEO8FSVXKpcaMhYUFKSEhATt2HH+ncA2m002m62qywAAoEr5+3oryh5Q7lyUPUDj+t1Q7tyhnFO6c2KaDpw4JdOUXv9um3q0iFZYjerzXljl4cPhcGjLli3q2LFjVT8VAAAuL8oeoNRnO6nElHqPX6aNB3LU+qUfVC8sSCq/XURBfj5q36CWnvjDdQr084w/z0hVED6efvpp9ezZU3Xq1FFWVpZeeukl5ebmatCgQZX9VAAAuCXDMORtSM/3aKL+H6xQUYlZ9qeY39t4IEfzN2VqxmMdFBrk5+RKq0alh4/9+/fr3nvvVXZ2tmrXrq2kpCStWLFCcXFxlf1UAAC4tbbxoVrzfGftP35SBY7ico+dOlOsf6/Zr9nrD2rv0ZP6f1+v10eD21hUaeWq9PAxbdq0yv6RAAB4LHuAr+wB52+77dggTDE1AzRh8U4t2palzJzTirT7O7nCysdnuwAA4KK8vAw907WR2tStqRJTmr5uv9UlVQrCBwAALu6u1rGSpK9X71cVT8hwCsIHAAAurntClAL9vLU7u0Br9h63upyrRvgAAMDFBdl8dFtClCRp8oq9OnDilLLyTut4QaHOFJdYXF3FeU7TMAAAHuzu1rH615r9mpl+UDPTD5adD7b5aNrQJDWNdp/PiuHOBwAAbqBN3Zr6U9MI1bD5yM/HS79+fl2eo0jT1x6wtrgK4s4HAABuwDAMvTewddnXpmlq9vqDemJaupZsP6LnLaytorjzAQCAGzIMQ7dcHy5vL0O/ZOVr//GTVpd02QgfAAC4KXuAr1rWuUaStHjbEWuLqQDCBwAAbuyW68MlET4AAICT3NywtiRp0bYsOYqKL3G1ayB8AADgxppGh6h2sE3FJabunJhmdTmXhfABAIAbMwxD97YpHb++8UCOMo65/sZTwgcAAG7uyS7X68azG09/3OH6ez8IHwAAeIBOZzeePjdjk1bvOWZxNRdH+AAAwAPclhBZ9s93vpumk4VFFlZzcYQPAAA8QIPwYE0a3Kbs63X7TlhXzCUQPgAA8BCdGoXr9hbRkqQ5Gw7pp11HdfDEKYurOhfhAwAAD9ImPlSSNHXlPvV7f4U6j1uiw7mnLa6qPD5YDgAAD9IjIUpzNxxUVp5Du44UqKCwWIkv/0cdrwuTJLWOC9UTf7zO0hoN0zRNSyv4ndzcXNntduXk5CgkJMTqcgAAcFsvz9ui93/cdc75r4e1U5u6oZX6XBV5/+bOBwAAHmrUH6/TjbHXyFFUUvr1l+mSpK9XZ1R6+KgIwgcAAB4q0M9H3RKiyr5evjNbX63er8KzYcQqbDgFAKCaSIi5RpLK7oRYhfABAEA1YfMpfdsnfAAAAKf4NXycPlNsaR2EDwAAqgmbj7ck7nwAAAAn+fXOBxtOAQCAU3h7GZKk4hJrR3wRPgAAqCZ8CB8AAMCZfr3zUVTCn10AAIAT+Hhz5wMAADiRt1fp234R4QMAADgDez4AAIBT/XfPB+EDAAA4AXc+AACAU5Xd+Sim2wUAADiBz9kNp9z5AAAATuHt7Rp7PnwsfXYAAOA0wf4+euyW+vL1tvbeA+EDAIBqIsTfV890bWR1GfzZBQAAOBfhAwAAOBXhAwAAOBXhAwAAOBXhAwAAOFWVhY8JEyYoPj5e/v7+atWqlZYuXVpVTwUAANxIlYSPL7/8UqNGjdJzzz2ndevWqWPHjurWrZv27dtXFU8HAADciGGaZqWPOUtMTFTLli01ceLEsnONGzdW7969NXbs2It+b25urux2u3JychQSElLZpQEAgCpQkffvSr/zUVhYqDVr1qhLly7lznfp0kXLly8/53qHw6Hc3NxyBwAA8FyVHj6ys7NVXFysiIiIcucjIiKUmZl5zvVjx46V3W4vO2JjYyu7JAAA4EKqbMOpYRjlvjZN85xzkjR69Gjl5OSUHRkZGVVVEgAAcAGV/tkuYWFh8vb2PucuR1ZW1jl3QyTJZrPJZrNVdhkAAMBFVfqdDz8/P7Vq1UoLFiwod37BggVq3759ZT8dAABwM1XyqbZPPvmkBg4cqNatW6tdu3Z6//33tW/fPg0bNuyS3/tr8w0bTwEAcB+/vm9fThNtlYSPfv366ejRo/rb3/6mQ4cOqVmzZpo3b57i4uIu+b15eXmSxMZTAADcUF5enux2+0WvqZI5H1ejpKREBw8eVHBw8Hk3qF6N3NxcxcbGKiMjgxkiVYh1dg7W2TlYZ+dgnateVa+xaZrKy8tTdHS0vLwuvqujSu58XA0vLy/FxMRU6XOEhITw4nYC1tk5WGfnYJ2dg3WuelW5xpe64/ErPlgOAAA4FeEDAAA4VbUKHzabTS+88AJzRaoY6+wcrLNzsM7OwTpXPVdaY5fbcAoAADxbtbrzAQAArEf4AAAATkX4AAAATkX4AAAATlVtwseECRMUHx8vf39/tWrVSkuXLrW6JLcyZswYGYZR7oiMjCx73DRNjRkzRtHR0QoICNAtt9yizZs3l/sZDodDI0eOVFhYmIKCgnT77bdr//79zv5VXMqPP/6onj17Kjo6WoZhaObMmeUer6x1PX78uAYOHCi73S673a6BAwfqxIkTVfzbuY5LrfPgwYPPeX0nJSWVu4Z1vrixY8eqTZs2Cg4OVnh4uHr37q1t27aVu4bX89W7nHV2h9dztQgfX375pUaNGqXnnntO69atU8eOHdWtWzft27fP6tLcStOmTXXo0KGyY+PGjWWPvfbaaxo3bpxSUlK0atUqRUZGqnPnzmWf1SNJo0aN0owZMzRt2jSlpqYqPz9fPXr0UHFxsRW/jksoKChQixYtlJKSct7HK2td+/fvr/T0dM2fP1/z589Xenq6Bg4cWOW/n6u41DpLUteuXcu9vufNm1fucdb54pYsWaLhw4drxYoVWrBggYqKitSlSxcVFBSUXcPr+epdzjpLbvB6NquBtm3bmsOGDSt3rlGjRub//M//WFSR+3nhhRfMFi1anPexkpISMzIy0nzllVfKzp0+fdq02+3mu+++a5qmaZ44ccL09fU1p02bVnbNgQMHTC8vL3P+/PlVWru7kGTOmDGj7OvKWteff/7ZlGSuWLGi7Jq0tDRTkrl169Yq/q1cz+/X2TRNc9CgQWavXr0u+D2sc8VlZWWZkswlS5aYpsnruar8fp1N0z1ezx5/56OwsFBr1qxRly5dyp3v0qWLli9fblFV7mnHjh2Kjo5WfHy87rnnHu3atUuStHv3bmVmZpZbY5vNpptvvrlsjdesWaMzZ86UuyY6OlrNmjXj38MFVNa6pqWlyW63KzExseyapKQk2e121v43Fi9erPDwcDVs2FBDhgxRVlZW2WOsc8Xl5ORIkkJDQyXxeq4qv1/nX7n669njw0d2draKi4sVERFR7nxERIQyMzMtqsr9JCYm6rPPPtN3332nDz74QJmZmWrfvr2OHj1ato4XW+PMzEz5+fmpZs2aF7wG5VXWumZmZio8PPycnx8eHs7an9WtWzd98cUXWrhwod58802tWrVKt956qxwOhyTWuaJM09STTz6p5ORkNWvWTBKv56pwvnWW3OP17HKfaltVDMMo97Vpmuecw4V169at7J8TEhLUrl071a9fX59++mnZRqYrWWP+PVxaZazr+a5n7f+rX79+Zf/crFkztW7dWnFxcZo7d6769u17we9jnc9vxIgR2rBhg1JTU895jNdz5bnQOrvD69nj73yEhYXJ29v7nKSWlZV1TgLH5QsKClJCQoJ27NhR1vVysTWOjIxUYWGhjh8/fsFrUF5lrWtkZKQOHz58zs8/cuQIa38BUVFRiouL044dOySxzhUxcuRIzZ49W4sWLVJMTEzZeV7PletC63w+rvh69vjw4efnp1atWmnBggXlzi9YsEDt27e3qCr353A4tGXLFkVFRSk+Pl6RkZHl1riwsFBLliwpW+NWrVrJ19e33DWHDh3Spk2b+PdwAZW1ru3atVNOTo5WrlxZds1PP/2knJwc1v4Cjh49qoyMDEVFRUlinS+HaZoaMWKEpk+froULFyo+Pr7c47yeK8el1vl8XPL1fNVbVt3AtGnTTF9fX/Ojjz4yf/75Z3PUqFFmUFCQuWfPHqtLcxtPPfWUuXjxYnPXrl3mihUrzB49epjBwcFla/jKK6+YdrvdnD59urlx40bz3nvvNaOioszc3NyynzFs2DAzJibG/OGHH8y1a9eat956q9miRQuzqKjIql/Lcnl5eea6devMdevWmZLMcePGmevWrTP37t1rmmblrWvXrl3N5s2bm2lpaWZaWpqZkJBg9ujRw+m/r1Uuts55eXnmU089ZS5fvtzcvXu3uWjRIrNdu3bmtddeyzpXwKOPPmra7XZz8eLF5qFDh8qOkydPll3D6/nqXWqd3eX1XC3Ch2ma5vjx4824uDjTz8/PbNmyZbm2JFxav379zKioKNPX19eMjo42+/bta27evLns8ZKSEvOFF14wIyMjTZvNZt50003mxo0by/2MU6dOmSNGjDBDQ0PNgIAAs0ePHua+ffuc/au4lEWLFpmSzjkGDRpkmmblrevRo0fNAQMGmMHBwWZwcLA5YMAA8/jx4076La13sXU+efKk2aVLF7N27dqmr6+vWadOHXPQoEHnrCHrfHHnW19J5qRJk8qu4fV89S61zu7yejbO/jIAAABO4fF7PgAAgGshfAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKcifAAAAKf6/2YTCErhQGnvAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "plt.plot(mros_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07305423-55a9-4d83-84f7-b51c07586587",
   "metadata": {},
   "source": [
    "**2. Selecting a threshold by maximizing the pairwise correlation between predicted outdegrees from multiple GRN inference methods**\n",
    "\n",
    "Here, you should calculate the average pairwise correlation between regulator outdegrees from different GRN predictions (previously described [here](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-03987-y)). We have provided a function \"avg_pairwise_correlation_outdegree\" which calculates the pairwise correlation between a set of GRNs for a specified number of interactions in the netẃorks. This function takes two parameters as input: 1) a dictionary with GRNs, 2) the number of interactions to use from the networks. \n",
    "\n",
    "Calculate the pairwise correlation for different number of interactions in the networks by changing the parameter network_cutoff. Lastly, you can plot the average pairwise correlation vs the network cutoff (using e.g. plot from matplotlib.pyplot).\n",
    "\n",
    "The optimal threshold is were the maximum average pairwise correlation is achieved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f59eee7-1329-47ae-a3e2-faf3e8758d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def avg_pairwise_correlation_outdegree(grns, network_cutoff):\n",
    "    # Count regulator outdegrees for a specified network cutoff\n",
    "    regulator_outdegrees = []\n",
    "    for grn_name in grns:\n",
    "        grn_subset = grns[grn_name].iloc[:network_cutoff, :]\n",
    "        ods = get_regulator_outdegree(grn_subset)\n",
    "        ods.columns = [grn_name]\n",
    "        regulator_outdegrees += [ods]\n",
    "    regulator_outdegrees = pd.concat(regulator_outdegrees, axis=1)\n",
    "    \n",
    "    # Calculate pairwis correlation\n",
    "    pair_corr = regulator_outdegrees.corr().values[np.triu_indices(len(grns.keys()), k=1)]\n",
    "    return pair_corr.mean()\n",
    "\n",
    "# Run function for a range of network cutoffs\n",
    "pair_corr = avg_pairwise_correlation_outdegree(grns, network_cutoff=10000)\n",
    "\n",
    "# Plot average pairwise correlation vs network cutoff\n",
    "# plt.plot(x, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4a55fbe-df8a-4343-a08c-cae6f76b6ebf",
   "metadata": {},
   "source": [
    "**3. Now you should select a network cutoff and change your networks accordingly.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b4a39a-9a26-4de9-bcac-f778f858ce12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Below is some example code.\n",
    "network_cutoff =\n",
    "grns['PCC'] = grn.iloc[:network_cutoff, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e08d524-ab75-4b89-91c9-b9572c5730d9",
   "metadata": {},
   "source": [
    "## 5. Consensus network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "859102ab-8dbe-40b2-8030-72038f47d79d",
   "metadata": {},
   "source": [
    "Many GRN inference methods have different performance on different datasets. Combining GRNs of multiple methods can result in more stable predictions across datasets. Therefore, using an ensemble approach can be a good idea if you are unsure of which method to apply. There are multiple approaches for making consensus networks. Two options are:\n",
    "1. Selecting interactions which are present in multiple inferred GRNs.\n",
    "2. Ranking interactions based on confidence scores and select the top interactions based on average ranking (described [here](https://www.nature.com/articles/nmeth.2016)).\n",
    "\n",
    "In this lab we suggest that you use the first approach to make a consensus network. You can make multiple networks by setting that interactions should be present in at least two or three GRN predictions. To your help we have provided a function \"create_edge_matrix\" which transforms a GRN in sparse format to an interaction matrix with 0 (no interaction) and 1 (interaction)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4386cc04-2908-49c6-9b44-4eda53597b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_edge_matrix(grn):\n",
    "    grn_mat = grn.pivot(columns='target', index='regulator', values='confidence').fillna(0)\n",
    "    grn_mat[grn_mat > 0] = 1\n",
    "    return grn_mat\n",
    "\n",
    "grn_mat = create_edge_matrix(grns['PCC']).reindex(index=regulators, columns=exp.columns).fillna(0)\n",
    "grn_mat\n",
    "\n",
    "# Make consensus network\n",
    "#consensus_grn =\n",
    "\n",
    "# Store consensus network in your GRN dictionary\n",
    "#grns['consensus'] = consensus_grn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fc8e23e-a187-4104-b680-7ce8be4ca798",
   "metadata": {},
   "source": [
    "## 6. Comparing topological properties\n",
    "In this section you will explore different topological properties of your GRNs. We will use networkx for this and a good first step is to create networkx Graphs with your GRNs. Below is some code which creates networkx graphs from a dictionary with GRNs and stores the GRN graphs in another dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f93aa540-adc0-4dbf-ad0d-a749f3e68039",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "\n",
    "grn_graphs = {}\n",
    "for grn_name, grn in grns.items():\n",
    "    grn_graphs[grn_name] = nx.from_pandas_edgelist(\n",
    "        grn, \n",
    "        source='regulator', \n",
    "        target='target',\n",
    "        create_using=nx.DiGraph)\n",
    "grn_graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad4e9355-6eb1-416b-97a1-a28f95fc273e",
   "metadata": {},
   "source": [
    "### Outdegree distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f930184-7299-4e7b-bafd-7faa103a38ec",
   "metadata": {},
   "source": [
    "Above we used pandas and the methods group_by and count to calculate outdegrees. Networkx has a specific method, \"out_degree\", for this purpose. This function calculates the outdegree of all nodes in the network but nodes which are not regulators will obtain an outdegree of 0. This function can be used as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77517c86-9a65-4b1d-8680-558174e46d31",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "grn_g = grn_graphs['PCC']\n",
    "dict(grn_g.out_degree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ea068d-b4db-4354-9ddf-599de9c5895e",
   "metadata": {},
   "source": [
    "What is the outdegree distribution of your predicted GRNs? First, calculate the outdegrees for all regulators and then plot a histogram to show the distribution of outdegrees. Below is some code that can be adjusted to plot the distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e5754a-18a6-4331-bda0-38a46d2b106e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Below is some example code.\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(ncols=len(grn_graphs.keys()), figsize=(12, 3))\n",
    "for i, grn_name in enumerate(grn_graphs):\n",
    "    # Make a list with all outdegrees here.\n",
    "    outdegrees = \n",
    "\n",
    "    # Plot histograms\n",
    "    ax[i].hist(outdegrees, bins=50)\n",
    "    ax[i].set_title(grn_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89a1436-d4c9-4b2e-8feb-61232a026948",
   "metadata": {},
   "source": [
    "The oudegree distribution of GRNs are generally assumed to follow power-law. Does the outdegree distribution of your inferred GRNs follow power-law?\n",
    "\n",
    "This can be checked by plotting the cumulative distribution vs the outdegree. If the log(cumulative distribution) vs log(outdegree) is a straight line the distribution is approximately power-law.\n",
    "\n",
    "The code below can help you with this problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "080eb23f",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_g = grn_graphs['PCC']\n",
    "\n",
    "# Sort outdegrees\n",
    "sorted_outdegree = np.sort(list(dict(grn_g.out_degree).values()))\n",
    "sorted_outdegree = sorted_outdegree[sorted_outdegree != 0][::-1]\n",
    "\n",
    "# Get cumulative distribution\n",
    "cumulative_distribution = np.arange(1, len(sorted_outdegree) + 1)\n",
    "\n",
    "# Plot cumulative distributin vs outdegree on log-log scale\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef375c57-132d-4738-a965-ef75a047c452",
   "metadata": {},
   "source": [
    "Lastly, you can check how well the regulator outdegrees correlate between different inferred GRNs. Below we have provided some code that will create a pandas DataFrame with regulator outdegrees for multiple GRNs and plot a heatmap with the outdegree correlations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c26ff32-4a1b-4a22-a3e2-8898845b216f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_regulator_outdegrees(grn_graphs):\n",
    "    regulator_outdegrees = []\n",
    "    for grn_name in grn_graphs:\n",
    "        regulator_outdegrees += [pd.Series(dict(grn_graphs[grn_name].out_degree)).rename(grn_name)]\n",
    "\n",
    "    regulator_outdegrees = pd.concat(regulator_outdegrees, axis=1).fillna(0)\n",
    "    return regulator_outdegrees\n",
    "\n",
    "regulator_outdegrees = get_regulator_outdegrees(grn_graphs)\n",
    "sns.heatmap(regulator_outdegrees.corr())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72b4c2e4-9634-4d79-9bf2-59a909b0f8f5",
   "metadata": {},
   "source": [
    "### Hubs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a2349e2-3e38-4fbd-a7db-2bf297ab0e83",
   "metadata": {},
   "source": [
    "A hub is a regulator which governs a significantly larger number of genes compared to the average regulator. Hubs are important for the structure of GRNs and the removal of a hub node tend to have large impact on the function of GRNs.\n",
    "\n",
    "The threshold for selecting a hub is often arbitrarily determined. However, a common approach is to select the top 10% regulators based on outdegree. \n",
    "\n",
    "You should identify the hubs in all of your inferred GRNs by for example selecting the top 10% regulators with the largest outdegree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a040970-1bd1-44c5-b0ad-4d2847076eb3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8537319c-bc12-4d11-b84f-8b971514a132",
   "metadata": {},
   "source": [
    "### Average shortest path length\n",
    "\n",
    "Find the average shortest path length of your inferred GRNs. Networkx has a function, \"shortest_path\", for obtaining the shortest path between two nodes. If no path exist between two nodes this function will fail. Usually the average path length is counted as 0 if no path exist between two nodes. You can stop the function from failing by for example using a try and except block. Make sure that the shortest path function works before using the try and except block."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c409c14a-8de0-47c2-a334-ab45fa9c0c6f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#try:\n",
    "shortest_paths = nx.shortest_path(grn_graphs['PCC'], \n",
    "                                  source=node1, \n",
    "                                  target=node2)\n",
    "#except:\n",
    "    #shortest_path_length = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f620d1e5-2c77-4b26-ad32-95029dde9d54",
   "metadata": {},
   "source": [
    "### Betweeness Centrality\n",
    "\n",
    "Calculate the betweeness centrality of all regulators in the network. This can be done with the \"betweenness_centrality\" function from networkx. Plot the betweeness centrality distribution of the GRNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2152a6b-7832-4ef7-90a1-0521659f7be8",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_betweenness_centrality = dict(nx.betweenness_centrality(grn_graphs['PCC']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35edbd34-08ac-4326-afef-d85553b0434e",
   "metadata": {},
   "source": [
    "### Closeness Centrality\n",
    "\n",
    "Calculate the closeness centrality of all regulators in the network. This can be done with the \"closeness_centrality\" function from networkx. Plot the closeness centrality distribution of the GRNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87d6ac5a-9a22-4153-bc9d-51215db5d280",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_closeness_centrality = dict(nx.closeness_centrality(grn_graphs['PCC']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e31a49-3a02-4160-b6e8-d45e131b26f2",
   "metadata": {},
   "source": [
    "## 7. Comparison with gold standard network\n",
    "In this section you will evaluate how close your predicted GRNs are to a gold standard. You should compare three different metrics:\n",
    "- Area under the precision recall curve (AUPR)\n",
    "- Outdegree correlation\n",
    "- Hub overlap\n",
    "\n",
    "Below are some code for obtaining the gold standard network. The gold standard is a signed network but the inferred GRNs for this lab only predict if an interactions exists or not. Therefore we will ignore the sign and only use the regulator and target columns of the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a903be6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "gold_standard = pd.read_csv(\"GRN_lab_signed_gold_standard.csv\")\n",
    "gold_standard = gold_standard.iloc[:, :2]\n",
    "gold_standard"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b62c3457",
   "metadata": {},
   "source": [
    "**AUPR**\n",
    "\n",
    "For this metric you need to use the confidence score matrices for each GRN. You can use the function average_precision_score(y_true, y_score) from sklearn.metrics to calculate the AUPR. y_true is a binary array with the gold standard interactions: 0 = no interaction and 1 = interaction. y_pred are all predicted interactions ranked on confidence score. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f552a469-a656-4013-b95e-c17b4f4bd6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import roc_auc_score, average_precision_score\n",
    "y_true =\n",
    "y_pred_rank =\n",
    "\n",
    "aupr = average_precision_score(y_true, y_pred_rank)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b999ff",
   "metadata": {},
   "source": [
    "**Outdegree correlation**\n",
    "\n",
    "Calculate the correlation between the regulator outdegrees for all inferred GRNs and gold standard."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "240a7cce-3e7c-4aac-a706-b890a95e400a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3d5ca536-0d59-4d06-b2da-7b23bfe4739b",
   "metadata": {},
   "source": [
    "**Hub  overlap**\n",
    "\n",
    "Identify the hubs in the gold standard and check if these overlap with the hubs in the predicted GRNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dddf5dd-8d29-4cd2-a5ac-8eb1dc541ec2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4e1155dd-25ea-4185-b550-161e9edcf543",
   "metadata": {},
   "source": [
    "## 8. Plotting GRNs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "229b8e74-15c9-4bd5-a7a6-050bf6dc1448",
   "metadata": {},
   "source": [
    "Below are some code for visualizing a directed network using networkx."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0cf2cf-7acf-4189-be84-62f578744d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "# Select network to plot\n",
    "grn_g = grn_graphs['PCC']\n",
    "\n",
    "# Color tfs with red and target genes with blue.\n",
    "node_color = ['r' if node in regulators else 'b' for node in list(grn_g.nodes())]\n",
    "\n",
    "# Adjust size of network by adjusting the figsize parameter.\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "# Draw network\n",
    "nx.draw_networkx(grn_g, ax=ax,\n",
    "                 pos=nx.spring_layout(grn_g),\n",
    "                 node_color=node_color, edge_color='grey')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46926596-2c42-4dd0-802a-517a74bd7464",
   "metadata": {},
   "source": [
    "You can subset nodes in the network to only plot certain parts. Below the most strongly connected component of the network is plotted. If you have time you can plot different subsets of nodes such as only plotting hubs and their target genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f4d1bf-057b-4235-aa98-4ffc3178f291",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only plot a subset od nodes.\n",
    "subset_nodes = max(nx.strongly_connected_components(grn_g), key=len)\n",
    "grn_subg = nx.subgraph(grn_g, subset_nodes)\n",
    "\n",
    "# Color tfs with red and target genes with blue.\n",
    "node_color = ['r' if node in regulators else 'b' for node in list(grn_subg.nodes())]\n",
    "\n",
    "# Adjust size of network by adjusting the figsize parameter.\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "\n",
    "# Draw network\n",
    "nx.draw_networkx(grn_subg, ax=ax,\n",
    "                 pos=nx.spring_layout(grn_subg),\n",
    "                 node_color=node_color, edge_color='grey')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a28436a-c4c7-4939-8a75-f55d073e0be1",
   "metadata": {},
   "source": [
    "## 9. Biological interpretation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78a2627c-dd33-488f-987b-4ed43fd4bf9f",
   "metadata": {},
   "source": [
    "There are a few databases which have gathered experimental and predicted interactions in GRNs. One of these are [TTRUST](https://www.grnpedia.org/trrust) which have gathered regulatory links for Human and Mouse. In this section you will explore the Human GRN from TTRUST and investigate how genes associated with multiple sclerosis (MS) are related to each other in the network. The MS genes are obtained from the [DisGeNet](https://www.disgenet.org/) database.\n",
    "\n",
    "Download the following files from the course Canvas page:\n",
    "- trrust_rawdata.human.tsv\n",
    "- MS_C0026769_disease_gda_summary.tsv\n",
    "\n",
    "You can start by exploring the following:\n",
    "1. How many regulators and target genes are in TTRUST?\n",
    "2. What is the degree distribution of the TTRUST GRN?\n",
    "3. Does the outdegree distibution of TTRUST follow power-law?\n",
    "4. What is the average outdegree of the regulators in TTRUST?\n",
    "5. Which are the hubs in TTRUST?\n",
    "6. Are any of the MS genes regulators? Any hubs?\n",
    "7. Which regulators are connected to MS genes. Do these regulators have a large or small outdegree? Any hubs?\n",
    "8. Do the MS genes have high betweeness centrality and/or closeness centrality.\n",
    "9. Are the MS genes close to each other in the network?\n",
    "\n",
    "Below is some code to obtain the TTRUST GRN and the MS genes. Make sure that you adjust the file paths to your file system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96888da0-093d-4bd7-85a2-c65f7df2aeb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Get human GRN from TTRUST\n",
    "ttrust_grn = pd.read_csv(\"trrust_rawdata.human.tsv\", sep='\\t', header=None).iloc[:, :2]\n",
    "print('GRN shape:', ttrust_grn.shape)\n",
    "\n",
    "# Get MS genes from DisGeNet with evidence score above 0.1\n",
    "dgn_ms = pd.read_csv(\"MS_C0026769_disease_gda_summary.tsv\", sep='\\t')\n",
    "dgn_ms = dgn_ms[(dgn_ms['Score_gda'] > 0.1)]['Gene'].unique()\n",
    "\n",
    "# MS genes present in TTRUST\n",
    "dgn_ms = np.intersect1d(dgn_ms, np.unique(ttrust_grn.values))\n",
    "print('Number of MS genes:', len(dgn_ms))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb84c287-3a0d-4953-b329-7165355606da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ef35db-392f-417f-8ba8-957ced6a309f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
